<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.313">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Stephen D. Turner, Ph.D.">
<title>Biological Data Science with R - 8&nbsp; Essential statistics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./survival.html" rel="next">
<link href="./rmarkdown.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script><script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script><script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./stats.html">Electives</a></li><li class="breadcrumb-item"><a href="./stats.html"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Essential statistics</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Biological Data Science with R</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/stephenturner/bdsr" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Biological-Data-Science-with-R.pdf">
              <i class="bi bi-bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Biological-Data-Science-with-R.epub">
              <i class="bi bi-bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Core Curriculum</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Basics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tibbles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Tibbles</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dplyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation with dplyr</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tidyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Tidy Data and Advanced Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ggplot2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Visualization with ggplot2</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tidyeda.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Refresher: Tidy Exploratory Data Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./rmarkdown.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Reproducible Reporting with RMarkdown</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Electives</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./stats.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Essential statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./survival.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Survival Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./predmodeling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Predictive Analytics: Predicting and Forecasting Influenza</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./textmining.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Text Mining and NLP</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./rnaseq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Count-Based Differential Expression Analysis of RNA-seq Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ggtree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Visualizing and Annotating Phylogenetic Trees</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Setup</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Further Resources</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li>
<a href="#our-data-nhanes" id="toc-our-data-nhanes" class="nav-link active" data-scroll-target="#our-data-nhanes"><span class="header-section-number">8.1</span> Our data: NHANES</a>
  <ul class="collapse">
<li><a href="#about-nhanes" id="toc-about-nhanes" class="nav-link" data-scroll-target="#about-nhanes"><span class="header-section-number">8.1.1</span> About NHANES</a></li>
  <li><a href="#import-inspect" id="toc-import-inspect" class="nav-link" data-scroll-target="#import-inspect"><span class="header-section-number">8.1.2</span> Import &amp; inspect</a></li>
  </ul>
</li>
  <li>
<a href="#descriptive-statistics" id="toc-descriptive-statistics" class="nav-link" data-scroll-target="#descriptive-statistics"><span class="header-section-number">8.2</span> Descriptive statistics</a>
  <ul class="collapse">
<li><a href="#missing-data" id="toc-missing-data" class="nav-link" data-scroll-target="#missing-data"><span class="header-section-number">8.2.1</span> Missing data</a></li>
  <li><a href="#eda" id="toc-eda" class="nav-link" data-scroll-target="#eda"><span class="header-section-number">8.2.2</span> EDA</a></li>
  </ul>
</li>
  <li>
<a href="#continuous-variables" id="toc-continuous-variables" class="nav-link" data-scroll-target="#continuous-variables"><span class="header-section-number">8.3</span> Continuous variables</a>
  <ul class="collapse">
<li><a href="#t-tests" id="toc-t-tests" class="nav-link" data-scroll-target="#t-tests"><span class="header-section-number">8.3.1</span> T-tests</a></li>
  <li><a href="#wilcoxon-test" id="toc-wilcoxon-test" class="nav-link" data-scroll-target="#wilcoxon-test"><span class="header-section-number">8.3.2</span> Wilcoxon test</a></li>
  <li><a href="#linear-models" id="toc-linear-models" class="nav-link" data-scroll-target="#linear-models"><span class="header-section-number">8.3.3</span> Linear models</a></li>
  <li><a href="#anova" id="toc-anova" class="nav-link" data-scroll-target="#anova"><span class="header-section-number">8.3.4</span> ANOVA</a></li>
  <li><a href="#linear-regression" id="toc-linear-regression" class="nav-link" data-scroll-target="#linear-regression"><span class="header-section-number">8.3.5</span> Linear regression</a></li>
  <li><a href="#multiple-regression" id="toc-multiple-regression" class="nav-link" data-scroll-target="#multiple-regression"><span class="header-section-number">8.3.6</span> Multiple regression</a></li>
  </ul>
</li>
  <li>
<a href="#discrete-variables" id="toc-discrete-variables" class="nav-link" data-scroll-target="#discrete-variables"><span class="header-section-number">8.4</span> Discrete variables</a>
  <ul class="collapse">
<li><a href="#contingency-tables" id="toc-contingency-tables" class="nav-link" data-scroll-target="#contingency-tables"><span class="header-section-number">8.4.1</span> Contingency tables</a></li>
  <li><a href="#logistic-regression" id="toc-logistic-regression" class="nav-link" data-scroll-target="#logistic-regression"><span class="header-section-number">8.4.2</span> Logistic regression</a></li>
  </ul>
</li>
  <li>
<a href="#power-sample-size" id="toc-power-sample-size" class="nav-link" data-scroll-target="#power-sample-size"><span class="header-section-number">8.5</span> Power &amp; sample size</a>
  <ul class="collapse">
<li><a href="#t-test-powern" id="toc-t-test-powern" class="nav-link" data-scroll-target="#t-test-powern"><span class="header-section-number">8.5.1</span> T-test power/N</a></li>
  <li><a href="#proportions-powern" id="toc-proportions-powern" class="nav-link" data-scroll-target="#proportions-powern"><span class="header-section-number">8.5.2</span> Proportions power/N</a></li>
  </ul>
</li>
  <li><a href="#tidying-models" id="toc-tidying-models" class="nav-link" data-scroll-target="#tidying-models"><span class="header-section-number">8.6</span> Tidying models</a></li>
  <li>
<a href="#additional-topics-recommended-reading" id="toc-additional-topics-recommended-reading" class="nav-link" data-scroll-target="#additional-topics-recommended-reading"><span class="header-section-number">8.7</span> Additional topics &amp; recommended reading</a>
  <ul class="collapse">
<li><a href="#batch-effects" id="toc-batch-effects" class="nav-link" data-scroll-target="#batch-effects"><span class="header-section-number">8.7.1</span> 1. Batch effects</a></li>
  <li><a href="#whats-my-n" id="toc-whats-my-n" class="nav-link" data-scroll-target="#whats-my-n"><span class="header-section-number">8.7.2</span> 2. What’s my <em>n</em>?</a></li>
  <li><a href="#technical-versus-biological-replicates" id="toc-technical-versus-biological-replicates" class="nav-link" data-scroll-target="#technical-versus-biological-replicates"><span class="header-section-number">8.7.3</span> 3. Technical versus biological replicates</a></li>
  </ul>
</li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-stats" class="quarto-section-identifier"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Essential statistics</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><p>This chapter provides hands-on instruction and exercises covering basic statistical analysis in R. This will cover descriptive statistics, <em>t</em>-tests, linear models, chi-square, clustering, dimensionality reduction, and resampling strategies. We will also cover methods for “tidying” model results for downstream visualization and summarization.</p>
<p><strong>Handouts</strong>: Download and print out these handouts and bring them to class:</p>
<ul>
<li><a href="handouts/r-stats-cheatsheet.pdf">Cheat sheet</a></li>
<li><a href="handouts/r-stats-exercises.pdf">Exercises handout</a></li>
</ul>
<section id="our-data-nhanes" class="level2" data-number="8.1"><h2 data-number="8.1" class="anchored" data-anchor-id="our-data-nhanes">
<span class="header-section-number">8.1</span> Our data: NHANES</h2>
<section id="about-nhanes" class="level3" data-number="8.1.1"><h3 data-number="8.1.1" class="anchored" data-anchor-id="about-nhanes">
<span class="header-section-number">8.1.1</span> About NHANES</h3>
<p>The data we’re going to work with comes from the National Health and Nutrition Examination Survey (NHANES) program at the CDC. You can read a lot more about NHANES on the <a href="http://www.cdc.gov/nchs/nhanes/">CDC’s website</a> or <a href="https://en.wikipedia.org/wiki/National_Health_and_Nutrition_Examination_Survey">Wikipedia</a>. NHANES is a research program designed to assess the health and nutritional status of adults and children in the United States. The survey is one of the only to combine both survey questions and physical examinations. It began in the 1960s and since 1999 examines a nationally representative sample of about 5,000 people each year. The NHANES interview includes demographic, socioeconomic, dietary, and health-related questions. The physical exam includes medical, dental, and physiological measurements, as well as several standard laboratory tests. NHANES is used to determine the prevalence of major diseases and risk factors for those diseases. NHANES data are also the basis for national standards for measurements like height, weight, and blood pressure. Data from this survey is used in epidemiology studies and health sciences research, which help develop public health policy, direct and design health programs and services, and expand the health knowledge for the Nation.</p>
<p>We are using a small slice of this data. We’re only using a handful of variables from the 2011-2012 survey years on about 5,000 individuals. The CDC uses a <a href="http://www.cdc.gov/nchs/data/series/sr_02/sr02_162.pdf">sampling strategy</a> to purposefully oversample certain subpopulations like racial minorities. Naive analysis of the original NHANES data can lead to mistaken conclusions because the percentages of people from each racial group in the data are different from general population. The 5,000 individuals here are resampled from the larger NHANES study population to undo these oversampling effects, so you can treat this as if it were a simple random sample from the American population.</p>
<p>You can download the data here: <a href="data/nhanes.csv"><strong>nhanes.csv</strong></a>. There’s also a data dictionary here: <a href="data/nhanes_dd.csv"><strong>nhanes_dd.csv</strong></a> that lists and describes each variable in our NHANES dataset. This table is copied below.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-2_494e0a02e80de0557904a957fcc3211f">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 16%">
<col style="width: 83%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">Variable</th>
<th style="text-align: left;">Definition</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>id</strong></td>
<td style="text-align: left;">A unique sample identifier</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Gender</strong></td>
<td style="text-align: left;">Gender (sex) of study participant coded as male or female</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Age</strong></td>
<td style="text-align: left;">Age in years at screening of study participant. Note: Subjects 80 years or older were recorded as 80.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Race</strong></td>
<td style="text-align: left;">Reported race of study participant, including non-Hispanic Asian category: Mexican, Hispanic, White, Black, Asian, or Other. Not availale for 2009-10.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Education</strong></td>
<td style="text-align: left;">Educational level of study participant Reported for participants aged 20 years or older. One of 8thGrade, 9-11thGrade, HighSchool, SomeCollege, or CollegeGrad.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>MaritalStatus</strong></td>
<td style="text-align: left;">Marital status of study participant. Reported for participants aged 20 years or older. One of Married, Widowed, Divorced, Separated, NeverMarried, or LivePartner (living with partner).</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>RelationshipStatus</strong></td>
<td style="text-align: left;">Simplification of MaritalStatus, coded as Committed if MaritalStatus is Married or LivePartner, and Single otherwise.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Insured</strong></td>
<td style="text-align: left;">Indicates whether the individual is covered by health insurance.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Income</strong></td>
<td style="text-align: left;">Numerical version of HHIncome derived from the middle income in each category</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Poverty</strong></td>
<td style="text-align: left;">A ratio of family income to poverty guidelines. Smaller numbers indicate more poverty</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>HomeRooms</strong></td>
<td style="text-align: left;">How many rooms are in home of study participant (counting kitchen but not bathroom). 13 rooms = 13 or more rooms.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>HomeOwn</strong></td>
<td style="text-align: left;">One of Home, Rent, or Other indicating whether the home of study participant or someone in their family is owned, rented or occupied by some other arrangement.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Work</strong></td>
<td style="text-align: left;">Indicates whether the individual is current working or not.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Weight</strong></td>
<td style="text-align: left;">Weight in kg</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Height</strong></td>
<td style="text-align: left;">Standing height in cm. Reported for participants aged 2 years or older.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>BMI</strong></td>
<td style="text-align: left;">Body mass index (weight/height2 in kg/m2). Reported for participants aged 2 years or older.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Pulse</strong></td>
<td style="text-align: left;">60 second pulse rate</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>BPSys</strong></td>
<td style="text-align: left;">Combined systolic blood pressure reading, following the procedure outlined for BPXSAR.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>BPDia</strong></td>
<td style="text-align: left;">Combined diastolic blood pressure reading, following the procedure outlined for BPXDAR.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Testosterone</strong></td>
<td style="text-align: left;">Testerone total (ng/dL). Reported for participants aged 6 years or older. Not available for 2009-2010.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>HDLChol</strong></td>
<td style="text-align: left;">Direct HDL cholesterol in mmol/L. Reported for participants aged 6 years or older.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>TotChol</strong></td>
<td style="text-align: left;">Total HDL cholesterol in mmol/L. Reported for participants aged 6 years or older.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Diabetes</strong></td>
<td style="text-align: left;">Study participant told by a doctor or health professional that they have diabetes. Reported for participants aged 1 year or older as Yes or No.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>DiabetesAge</strong></td>
<td style="text-align: left;">Age of study participant when first told they had diabetes. Reported for participants aged 1 year or older.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>nPregnancies</strong></td>
<td style="text-align: left;">How many times participant has been pregnant. Reported for female participants aged 20 years or older.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>nBabies</strong></td>
<td style="text-align: left;">How many of participants deliveries resulted in live births. Reported for female participants aged 20 years or older.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>SleepHrsNight</strong></td>
<td style="text-align: left;">Self-reported number of hours study participant usually gets at night on weekdays or workdays. Reported for participants aged 16 years and older.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>PhysActive</strong></td>
<td style="text-align: left;">Participant does moderate or vigorous-intensity sports, fitness or recreational activities (Yes or No). Reported for participants 12 years or older.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>PhysActiveDays</strong></td>
<td style="text-align: left;">Number of days in a typical week that participant does moderate or vigorous-intensity activity. Reported for participants 12 years or older.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>AlcoholDay</strong></td>
<td style="text-align: left;">Average number of drinks consumed on days that participant drank alcoholic beverages. Reported for participants aged 18 years or older.</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>AlcoholYear</strong></td>
<td style="text-align: left;">Estimated number of days over the past year that participant drank alcoholic beverages. Reported for participants aged 18 years or older.</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>SmokingStatus</strong></td>
<td style="text-align: left;">Smoking status: Current Former or Never.</td>
</tr>
</tbody>
</table>
</div>
</div>
</section><section id="import-inspect" class="level3" data-number="8.1.2"><h3 data-number="8.1.2" class="anchored" data-anchor-id="import-inspect">
<span class="header-section-number">8.1.2</span> Import &amp; inspect</h3>
<p>First, let’s load the dplyr and readr libraries.</p>
<div class="cell" data-hash="stats_cache/html/loadpkgs_58031733c699c7c78a818f77557c4bf3">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you see a warning that looks like this: <code>Error in library(dplyr) : there is no package called 'dplyr'</code> (or similar with readr), then you don’t have the package installed correctly. See the (<a href="setup.html">Appendix&nbsp;<span>A</span></a>)</p>
<p>Now, let’s actually load the data. When we load data we assign it to a variable just like any other, and we can choose a name for that data. Since we’re going to be referring to this data a lot, let’s give it a short easy name to type. I’m going to call it <code>nh</code>. Once we’ve loaded it we can type the name of the object itself (<code>nh</code>) to see it printed to the screen.</p>
<div class="cell" data-hash="stats_cache/html/loaddata_1363afe7fc787d5f3eec94cb48c36288">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">nh</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"data/nhanes.csv"</span><span class="op">)</span></span>
<span><span class="va">nh</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5,000 × 32
      id Gender   Age Race    Education MaritalStatus RelationshipStatus Insured
   &lt;dbl&gt; &lt;chr&gt;  &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;     &lt;chr&gt;         &lt;chr&gt;              &lt;chr&gt;  
 1 62163 male      14 Asian   &lt;NA&gt;      &lt;NA&gt;          &lt;NA&gt;               Yes    
 2 62172 female    43 Black   High Sch… NeverMarried  Single             Yes    
 3 62174 male      80 White   College … Married       Committed          Yes    
 4 62174 male      80 White   College … Married       Committed          Yes    
 5 62175 male       5 White   &lt;NA&gt;      &lt;NA&gt;          &lt;NA&gt;               Yes    
 6 62176 female    34 White   College … Married       Committed          Yes    
 7 62178 male      80 White   High Sch… Widowed       Single             Yes    
 8 62180 male      35 White   College … Married       Committed          Yes    
 9 62186 female    17 Black   &lt;NA&gt;      &lt;NA&gt;          &lt;NA&gt;               Yes    
10 62190 female    15 Mexican &lt;NA&gt;      &lt;NA&gt;          &lt;NA&gt;               Yes    
# ℹ 4,990 more rows
# ℹ 24 more variables: Income &lt;dbl&gt;, Poverty &lt;dbl&gt;, HomeRooms &lt;dbl&gt;,
#   HomeOwn &lt;chr&gt;, Work &lt;chr&gt;, Weight &lt;dbl&gt;, Height &lt;dbl&gt;, BMI &lt;dbl&gt;,
#   Pulse &lt;dbl&gt;, BPSys &lt;dbl&gt;, BPDia &lt;dbl&gt;, Testosterone &lt;dbl&gt;, HDLChol &lt;dbl&gt;,
#   TotChol &lt;dbl&gt;, Diabetes &lt;chr&gt;, DiabetesAge &lt;dbl&gt;, nPregnancies &lt;dbl&gt;,
#   nBabies &lt;dbl&gt;, SleepHrsNight &lt;dbl&gt;, PhysActive &lt;chr&gt;, PhysActiveDays &lt;dbl&gt;,
#   AlcoholDay &lt;dbl&gt;, AlcoholYear &lt;dbl&gt;, SmokingStatus &lt;chr&gt;</code></pre>
</div>
</div>
<p>Take a look at that output. The nice thing about loading dplyr and reading data with readr functions is that data are displayed in a much more friendly way. This dataset has 5,000 rows and 32 columns. When you import/convert data this way and try to display the object in the console, instead of trying to display all 5,000 rows, you’ll only see about 10 by default. Also, if you have so many columns that the data would wrap off the edge of your screen, those columns will not be displayed, but you’ll see at the bottom of the output which, if any, columns were hidden from view.</p>
<blockquote class="blockquote">
<p><em><strong>A note on characters versus factors:</strong></em> One thing that you immediately notice is that all the categorical variables are read in as <em>character</em> data types. This data type is used for storing strings of text, for example, IDs, names, descriptive text, etc. There’s another related data type called <em><strong>factors</strong></em>. Factor variables are used to represent categorical variables with two or more <em>levels</em>, e.g., “male” or “female” for Gender, or “Single” versus “Committed” for RelationshipStatus. For the most part, statistical analysis treats these two data types the same. It’s often easier to leave categorical variables as characters. However, in some cases you may get a warning message alerting you that a character variable was converted into a factor variable during analysis. Generally, these warnings are nothing to worry about. You can, if you like, convert individual variables to factor variables, or simply use dplyr’s <code>mutate_if</code> to convert all character vectors to factor variables:</p>
</blockquote>
<div class="cell" data-hash="stats_cache/html/asfactor_73792df7b04f8ce74799bbffec1baac3">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">nh</span> <span class="op">&lt;-</span> <span class="va">nh</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate_all.html">mutate_if</a></span><span class="op">(</span><span class="va">is.character</span>, <span class="va">as.factor</span><span class="op">)</span></span>
<span><span class="va">nh</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now just take a look at just a few columns that are now factors. Remember, you can look at individual variables with the <code>mydataframe$specificVariable</code> syntax.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-3_5be0f704291455fefa613276221f599b">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">nh</span><span class="op">$</span><span class="va">RelationshipStatus</span></span>
<span><span class="va">nh</span><span class="op">$</span><span class="va">Race</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/levels.html">levels</a></span><span class="op">(</span><span class="va">nh</span><span class="op">$</span><span class="va">Race</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you want to see the whole dataset, there are two ways to do this. First, you can click on the name of the data.frame in the <strong>Environment</strong> panel in RStudio. Or you could use the <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code> function (<em>with a capital V</em>).</p>
<div class="cell" data-hash="stats_cache/html/view_ef5990d610d6d56f35c8a428ba3303bb">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="va">nh</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Recall several built-in functions that are useful for working with data frames.</p>
<ul>
<li>Content:
<ul>
<li>
<code><a href="https://rdrr.io/r/utils/head.html">head()</a></code>: shows the first few rows</li>
<li>
<code><a href="https://rdrr.io/r/utils/head.html">tail()</a></code>: shows the last few rows</li>
</ul>
</li>
<li>Size:
<ul>
<li>
<code><a href="https://rdrr.io/r/base/dim.html">dim()</a></code>: returns a 2-element vector with the number of rows in the first element, and the number of columns as the second element (the dimensions of the object)</li>
<li>
<code><a href="https://rdrr.io/r/base/nrow.html">nrow()</a></code>: returns the number of rows</li>
<li>
<code><a href="https://rdrr.io/r/base/nrow.html">ncol()</a></code>: returns the number of columns</li>
</ul>
</li>
<li>Summary:
<ul>
<li>
<code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> (or just <code><a href="https://rdrr.io/r/base/names.html">names()</a></code>): returns the column names</li>
<li>
<code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code> (from <strong>dplyr</strong>): Returns a glimpse of your data, telling you the structure of the dataset and information about the class, length and content of each column</li>
</ul>
</li>
</ul>
<div class="cell" data-hash="stats_cache/html/tibble_functions_8047ab0c49575cd121c7323315710f01">
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">nh</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">nh</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">nh</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">nh</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">nh</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="descriptive-statistics" class="level2" data-number="8.2"><h2 data-number="8.2" class="anchored" data-anchor-id="descriptive-statistics">
<span class="header-section-number">8.2</span> Descriptive statistics</h2>
<p>We can access individual variables within a data frame using the <code>$</code> operator, e.g., <code>mydataframe$specificVariable</code>. Let’s print out all the <strong><code>Race</code></strong> values in the data. Let’s then see what are the <code>unique</code> values of each. Then let’s calculate the <code>mean</code>, <code>median</code>, and <code>range</code> of the <strong><code>Age</code></strong> variable.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-4_27593904e85132741ad5ade5f682ad5a">
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Display all Race values</span></span>
<span><span class="va">nh</span><span class="op">$</span><span class="va">Race</span></span>
<span></span>
<span><span class="co"># Get the unique values of Race</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">nh</span><span class="op">$</span><span class="va">Race</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="va">nh</span><span class="op">$</span><span class="va">Race</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Do the same thing the dplyr way</span></span>
<span><span class="va">nh</span><span class="op">$</span><span class="va">Race</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">nh</span><span class="op">$</span><span class="va">Race</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html">unique</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Age mean, median, range</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">nh</span><span class="op">$</span><span class="va">Age</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/median.html">median</a></span><span class="op">(</span><span class="va">nh</span><span class="op">$</span><span class="va">Age</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/range.html">range</a></span><span class="op">(</span><span class="va">nh</span><span class="op">$</span><span class="va">Age</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You could also do the last few operations using dplyr, but remember, this returns a single-row, single-column tibble, <em>not</em> a single scalar value like the above. This is only really useful in the context of grouping and summarizing.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-5_a9d31d2c1ffd4fc7e974a2eb1cf810e0">
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Compute the mean age</span></span>
<span><span class="va">nh</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">Age</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Now grouped by other variables</span></span>
<span><span class="va">nh</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">Gender</span>, <span class="va">Race</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">Age</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> function (note, this is different from <strong>dplyr</strong>’s <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize()</a></code>) works differently depending on which kind of object you pass to it. If you run <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> on a data frame, you get some very basic summary statistics on each variable in the data.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-6_f634360505bf74458c88780cfde041f3">
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">nh</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="missing-data" class="level3" data-number="8.2.1"><h3 data-number="8.2.1" class="anchored" data-anchor-id="missing-data">
<span class="header-section-number">8.2.1</span> Missing data</h3>
<p>Let’s try taking the mean of a different variable, either the dplyr way or the simpler <code>$</code> way.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-7_74011f534ac084a49972a04afd088f84">
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># the dplyr way: returns a single-row single-column tibble/dataframe</span></span>
<span><span class="va">nh</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">Income</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># returns a single value</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">nh</span><span class="op">$</span><span class="va">Income</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>What happened there? <code>NA</code> indicates <em>missing data</em>. Take a look at the Income variable.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-8_8c84c42d1bdd966c1b760d34d752d4b4">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Look at just the Income variable</span></span>
<span><span class="va">nh</span><span class="op">$</span><span class="va">Income</span></span>
<span></span>
<span><span class="co"># Or view the dataset</span></span>
<span><span class="co"># View(nh)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Notice that there are lots of missing values for Income. Trying to get the mean a bunch of observations with some missing data returns a missing value by default. This is almost universally the case with all summary statistics – a single <code>NA</code> will cause the summary to return <code>NA</code>. Now look at the help for <code><a href="https://rdrr.io/r/base/mean.html">?mean</a></code>. Notice the <code>na.rm</code> argument. This is a logical (i.e., <code>TRUE</code> or <code>FALSE</code>) value indicating whether or not missing values should be removed prior to computing the mean. By default, it’s set to <code>FALSE</code>. Now try it again.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-9_1359d602fdbddcbe93419c406056cd11">
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">nh</span><span class="op">$</span><span class="va">Income</span>, na.rm<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 57077.66</code></pre>
</div>
</div>
<p>The <code><a href="https://rdrr.io/r/base/NA.html">is.na()</a></code> function tells you if a value is missing. Get the <code><a href="https://rdrr.io/r/base/sum.html">sum()</a></code> of that vector, which adds up all the <code>TRUE</code>s to tell you how many of the values are missing.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-10_8108d814cbda89104c8f18513f162aa0">
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">nh</span><span class="op">$</span><span class="va">Income</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html">is.na</a></span><span class="op">(</span><span class="va">nh</span><span class="op">$</span><span class="va">Income</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s talk about exploratory data analysis (EDA).</p>
</section><section id="eda" class="level3" data-number="8.2.2"><h3 data-number="8.2.2" class="anchored" data-anchor-id="eda">
<span class="header-section-number">8.2.2</span> EDA</h3>
<p>It’s always worth examining your data visually before you start any statistical analysis or hypothesis testing. We could spend an entire day on <strong><a href="https://en.wikipedia.org/wiki/Exploratory_data_analysis">exploratory data analysis</a></strong>. The data visualization section (<a href="ggplot2.html">Chapter&nbsp;<span>5</span></a>) covers this in much broader detail. Here we’ll just mention a few of the big ones: <strong>histograms</strong> and <strong>scatterplots</strong>.</p>
<section id="histograms" class="level4" data-number="8.2.2.1"><h4 data-number="8.2.2.1" class="anchored" data-anchor-id="histograms">
<span class="header-section-number">8.2.2.1</span> Histograms</h4>
<p>We can learn a lot from the data just looking at the value distributions of particular variables. Let’s make some histograms with ggplot2. Looking at BMI shows a few extreme outliers. Looking at weight initially shows us that the units are probably in kg. Replotting that in lbs with more bins shows a clear bimodal distribution. Are there kids in this data? The age distribution shows us the answer is <em>yes</em>.</p>
<div class="cell" data-hash="stats_cache/html/histograms_b524a6a4b330ec6d0a8373fd21c723b9">
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">nh</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">BMI</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>bins<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/histograms-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">nh</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">Weight</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>bins<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/histograms-2.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># In pounds, more bins</span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">nh</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">Weight</span><span class="op">*</span><span class="fl">2.2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>bins<span class="op">=</span><span class="fl">80</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/histograms-3.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">nh</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">Age</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span>bins<span class="op">=</span><span class="fl">30</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/histograms-4.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="scatterplots" class="level4" data-number="8.2.2.2"><h4 data-number="8.2.2.2" class="anchored" data-anchor-id="scatterplots">
<span class="header-section-number">8.2.2.2</span> Scatterplots</h4>
<p>Let’s look at how a few different variables relate to each other. E.g., height and weight:</p>
<div class="cell" data-hash="stats_cache/html/scatter_heightweight_2f3f1fa5a2ac3e30e461aa3d17db7b92">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">nh</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">Height</span>, <span class="va">Weight</span>, col<span class="op">=</span><span class="va">Gender</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/scatter_heightweight-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Let’s filter out all the kids, draw trend lines using a linear model:</p>
<div class="cell" data-hash="stats_cache/html/scatter_heightweight_colgender_d13b634b8e4d104ba8ed5ff028163cbc">
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">nh</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Age</span><span class="op">&gt;=</span><span class="fl">18</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">Height</span>, <span class="va">Weight</span>, col<span class="op">=</span><span class="va">Gender</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"lm"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/scatter_heightweight_colgender-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Check out the data visualization section (<a href="ggplot2.html">Chapter&nbsp;<span>5</span></a>) for much more on this topic.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 1
</div>
</div>
<div class="callout-body-container callout-body">
<p>What’s the mean 60-second pulse rate for all participants in the data?</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-12_3345d3d5e72db1871b06d945e8f58518">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 73.63382</code></pre>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 2
</div>
</div>
<div class="callout-body-container callout-body">
<p>What’s the range of values for diastolic blood pressure in all participants? (Hint: see help for <code><a href="https://rdrr.io/r/base/Extremes.html">min()</a></code>, <code><a href="https://rdrr.io/r/base/Extremes.html">max()</a></code>, and <code><a href="https://rdrr.io/r/base/range.html">range()</a></code> functions, e.g., enter <code><a href="https://rdrr.io/r/base/range.html">?range</a></code> without the parentheses to get help).</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-13_e58406b04454bfa4adfd7b5c81b62738">
<div class="cell-output cell-output-stdout">
<pre><code>[1]   0 116</code></pre>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 3
</div>
</div>
<div class="callout-body-container callout-body">
<p>What are the median, lower, and upper quartiles for the age of all participants? (Hint: see help for <code>median</code>, or better yet, <code>quantile</code>).</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-14_05209125fa043128653291c286ca4289">
<div class="cell-output cell-output-stdout">
<pre><code>  0%  25%  50%  75% 100% 
   0   17   36   54   80 </code></pre>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 4
</div>
</div>
<div class="callout-body-container callout-body">
<p>What’s the variance and standard deviation for income among all participants?</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-15_8a698067a635ea9bb8bb4f8fe749074e">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1121564068</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 33489.76</code></pre>
</div>
</div>
</div>
</div>
</section></section></section><section id="continuous-variables" class="level2" data-number="8.3"><h2 data-number="8.3" class="anchored" data-anchor-id="continuous-variables">
<span class="header-section-number">8.3</span> Continuous variables</h2>
<section id="t-tests" class="level3" data-number="8.3.1"><h3 data-number="8.3.1" class="anchored" data-anchor-id="t-tests">
<span class="header-section-number">8.3.1</span> T-tests</h3>
<p>First let’s create a new dataset from <code>nh</code> called <code>nha</code> that only has adults. To prevent us from making any mistakes downstream, let’s remove the <code>nh</code> object.</p>
<div class="cell" data-hash="stats_cache/html/filterAdults_99d109a97f85d82363aaeac45a079ff4">
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">nha</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">nh</span>, <span class="va">Age</span><span class="op">&gt;=</span><span class="fl">18</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/rm.html">rm</a></span><span class="op">(</span><span class="va">nh</span><span class="op">)</span></span>
<span><span class="co"># View(nha)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let’s do a few two-sample t-tests to test for <em>differences in means between two groups</em>. The function for a t-test is <code><a href="https://rdrr.io/r/stats/t.test.html">t.test()</a></code>. See the help for <code><a href="https://rdrr.io/r/stats/t.test.html">?t.test</a></code>. We’ll be using the <em>forumla</em> method. The usage is <strong><code>t.test(response~group, data=myDataFrame)</code></strong>.</p>
<ol type="1">
<li>Are there differences in age for males versus females in this dataset?</li>
<li>Does BMI differ between diabetics and non-diabetics?</li>
<li>Do single or married/cohabitating people drink more alcohol? Is this relationship significant?</li>
</ol>
<div class="cell" data-hash="stats_cache/html/ttests_55d5373dcf8ee7a099abf73787a23ebd">
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">Age</span><span class="op">~</span><span class="va">Gender</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  Age by Gender
t = 1.9122, df = 3697.2, p-value = 0.05593
alternative hypothesis: true difference in means between group female and group male is not equal to 0
95 percent confidence interval:
 -0.02776814  2.22191122
sample estimates:
mean in group female   mean in group male 
            47.06412             45.96704 </code></pre>
</div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">BMI</span><span class="op">~</span><span class="va">Diabetes</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  BMI by Diabetes
t = -11.379, df = 407.31, p-value &lt; 2.2e-16
alternative hypothesis: true difference in means between group No and group Yes is not equal to 0
95 percent confidence interval:
 -5.563596 -3.924435
sample estimates:
 mean in group No mean in group Yes 
         28.08753          32.83155 </code></pre>
</div>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">AlcoholYear</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  AlcoholYear by RelationshipStatus
t = 5.4315, df = 2674.8, p-value = 6.09e-08
alternative hypothesis: true difference in means between group Committed and group Single is not equal to 0
95 percent confidence interval:
 13.05949 27.81603
sample estimates:
mean in group Committed    mean in group Single 
               83.93416                63.49640 </code></pre>
</div>
</div>
<p>See the heading, <em>Welch Two Sample t-test</em>, and notice that the degrees of freedom might not be what we expected based on our sample size. Now look at the help for <code><a href="https://rdrr.io/r/stats/t.test.html">?t.test</a></code> again, and look at the <code>var.equal</code> argument, which is by default set to <code>FALSE</code>. One of the assumptions of the t-test is <a href="https://en.wikipedia.org/wiki/Homoscedasticity">homoscedasticity</a>, or homogeneity of variance. This assumes that the variance in the outcome (e.g., BMI) is identical across both levels of the predictor (diabetic vs non-diabetic). Since this is rarely the case, the t-test defaults to using the <a href="https://en.wikipedia.org/wiki/Welch%27s_t-test">Welch correction</a>, which is a more reliable version of the t-test when the homoscedasticity assumption is violated.</p>
<blockquote class="blockquote">
<p><i class="fa fa-exclamation-triangle" aria-hidden="true"></i> <em><strong>A note on one-tailed versus two-tailed tests:</strong></em> A two-tailed test is almost always more appropriate. The hypothesis you’re testing here is spelled out in the results (“alternative hypothesis: true difference in means is not equal to 0”). If the p-value is very low, you can reject the null hypothesis that there’s no difference in means. Because you typically don’t know <em>a priori</em> whether the difference in means will be positive or negative (e.g., we don’t know <em>a priori</em> whether Single people would be expected to drink more or less than those in a committed relationship), we want to do the two-tailed test. However, if we <em>only</em> wanted to test a very specific directionality of effect, we could use a one-tailed test and specify which direction we expect. This is more powerful if we “get it right”, but much less powerful for the opposite effect. Notice how the p-value changes depending on how we specify the hypothesis. Again, the <strong>two-tailed test is almost always more appropriate</strong>.</p>
</blockquote>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-16_e10b854927d091b8614882271df694e3">
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Two tailed</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">AlcoholYear</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Difference in means is &gt;0 (committed drink more)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">AlcoholYear</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span>, alternative<span class="op">=</span><span class="st">"greater"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Difference in means is &lt;0 (committed drink less)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">AlcoholYear</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span>, alternative<span class="op">=</span><span class="st">"less"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<blockquote class="blockquote">
<p><i class="fa fa-exclamation-triangle" aria-hidden="true"></i> <em><strong>A note on paired versus unpaired t-tests:</strong></em> The t-test we performed here was an unpaired test. Here the males and females are different people. The diabetics and nondiabetics are different samples. The single and committed individuals are completely independent, separate observations. In this case, an <em>unpaired</em> test is appropriate. An alternative design might be when data is derived from samples who have been measured at two different time points or locations, e.g., before versus after treatment, left versus right hand, etc. In this case, a <em><strong>paired t-test</strong></em> would be more appropriate. A paired test takes into consideration the intra and inter-subject variability, and is more powerful than the unpaired test. See the help for <code><a href="https://rdrr.io/r/stats/t.test.html">?t.test</a></code> for more information on how to do this.</p>
</blockquote>
</section><section id="wilcoxon-test" class="level3" data-number="8.3.2"><h3 data-number="8.3.2" class="anchored" data-anchor-id="wilcoxon-test">
<span class="header-section-number">8.3.2</span> Wilcoxon test</h3>
<p>Another assumption of the t-test is that data is normally distributed. Looking at the histogram for AlcoholYear shows that this data clearly isn’t.</p>
<div class="cell" data-hash="stats_cache/html/alcoholhist_1b31ac1bda7dd40ac5018935161cada4">
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">nha</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">AlcoholYear</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html">geom_histogram</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/alcoholhist-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The <a href="https://en.wikipedia.org/wiki/Mann%E2%80%93Whitney_U_test">Wilcoxon rank-sum test (a.k.a. Mann-Whitney <em>U</em> test)</a> is a nonparametric test of differences in mean that does not require normally distributed data. When data is perfectly normal, the t-test is uniformly more powerful. But when this assumption is violated, the t-test is unreliable. This test is called in a similar way as the t-test.</p>
<div class="cell" data-hash="stats_cache/html/wilcoxtest_757ba36cb39e739edbccc7df9c4564fa">
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html">wilcox.test</a></span><span class="op">(</span><span class="va">AlcoholYear</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum test with continuity correction

data:  AlcoholYear by RelationshipStatus
W = 1067954, p-value = 0.0001659
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
</div>
<p>The results are still significant, but much less than the p-value reported for the (incorrect) t-test above. Also note in the help for <code><a href="https://rdrr.io/r/stats/wilcox.test.html">?wilcox.test</a></code> that there’s a <code>paired</code> option here too.</p>
</section><section id="linear-models" class="level3" data-number="8.3.3"><h3 data-number="8.3.3" class="anchored" data-anchor-id="linear-models">
<span class="header-section-number">8.3.3</span> Linear models</h3>
<blockquote class="blockquote">
<p>Analysis of variance and linear modeling are complex topics that deserve an entire semester dedicated to theory, design, and interpretation. A very good resource is <a href="https://www.amazon.com/Introduction-Statistical-Learning-Applications-Statistics/dp/1461471370/ref=sr_1_1?ie=UTF8&amp;qid=1473087847&amp;sr=8-1&amp;keywords=introduction+statistical+learning&amp;tag=gettgenedone-20"><em>An Introduction to Statistical Learning: with Applications in R</em></a> by Gareth James, Daniela Witten, Trevor Hastie and Robert Tibshirani. The <a href="http://www-bcf.usc.edu/~gareth/ISL/ISLR%20Sixth%20Printing.pdf">PDF</a> of the book and all the R code used throughout are <a href="http://www-bcf.usc.edu/~gareth/ISL/">available <strong>free</strong> on the author’s website</a>. What follows is a necessary over-simplification with more focus on implementation, and less on theory and design.</p>
</blockquote>
<p>Where t-tests and their nonparametric substitutes are used for assessing the differences in means between two groups, ANOVA is used to assess the significance of differences in means between multiple groups. In fact, a t-test is just a specific case of ANOVA when you only have two groups. And both t-tests and ANOVA are just specific cases of linear regression, where you’re trying to fit a model describing how a continuous outcome (e.g., BMI) changes with some predictor variable (e.g., diabetic status, race, age, etc.). The distinction is largely semantic – with a linear model you’re asking, “do levels of a categorical variable affect the response?” where with ANOVA or t-tests you’re asking, “does the mean response differ between levels of a categorical variable?”</p>
<p>Let’s examine the relationship between BMI and relationship status (<code>RelationshipStatus</code> was derived from <code>MaritalStatus</code>, coded as <em>Committed</em> if MaritalStatus is Married or LivePartner, and <em>Single</em> otherwise). Let’s first do this with a t-test, and for now, let’s assume that the variances between groups <em>are</em> equal.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-17_0a4ca0df2daefa1ea05246718d9d5cfc">
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">BMI</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span>, var.equal<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Two Sample t-test

data:  BMI by RelationshipStatus
t = -1.5319, df = 3552, p-value = 0.1256
alternative hypothesis: true difference in means between group Committed and group Single is not equal to 0
95 percent confidence interval:
 -0.77817842  0.09552936
sample estimates:
mean in group Committed    mean in group Single 
               28.51343                28.85475 </code></pre>
</div>
</div>
<p>It looks like single people have a very slightly higher BMI than those in a committed relationship, but the magnitude of the difference is trivial, and the difference is not significant. Now, let’s do the same test in a linear modeling framework. First, let’s create the fitted model and store it in an object called <code>fit</code>.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-18_efcaa9916fec7ffc469652ef07feae77">
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">BMI</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>You can display the object itself, but that isn’t too interesting. You can get the more familiar ANOVA table by calling the <code><a href="https://rdrr.io/r/stats/anova.html">anova()</a></code> function on the <code>fit</code> object. More generally, the <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> function on a linear model object will tell you much more. (Note this is different from dplyr’s <strong>summarize</strong> function).</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-19_f62cc1326cd75bbd0d835e722dbca856">
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fit</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = BMI ~ RelationshipStatus, data = nha)

Coefficients:
             (Intercept)  RelationshipStatusSingle  
                 28.5134                    0.3413  </code></pre>
</div>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: BMI
                     Df Sum Sq Mean Sq F value Pr(&gt;F)
RelationshipStatus    1     98  98.320  2.3467 0.1256
Residuals          3552 148819  41.897               </code></pre>
</div>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = BMI ~ RelationshipStatus, data = nha)

Residuals:
    Min      1Q  Median      3Q     Max 
-12.813  -4.613  -0.955   3.287  52.087 

Coefficients:
                         Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)               28.5134     0.1388 205.440   &lt;2e-16
RelationshipStatusSingle   0.3413     0.2228   1.532    0.126

Residual standard error: 6.473 on 3552 degrees of freedom
  (153 observations deleted due to missingness)
Multiple R-squared:  0.0006602, Adjusted R-squared:  0.0003789 
F-statistic: 2.347 on 1 and 3552 DF,  p-value: 0.1256</code></pre>
</div>
</div>
<p>Go back and re-run the t-test assuming equal variances as we did before. Now notice a few things:</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-20_5590a353c4e2e7d2311bc07d02356baf">
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">BMI</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span>, var.equal<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol type="1">
<li>The p-values from all three tests (t-test, ANOVA, and linear regression) are all identical (p=0.1256). This is because they’re all identical: a t-test is a specific case of ANOVA, which is a specific case of linear regression. There may be some rounding error, but we’ll talk about extracting the exact values from a model object later on.</li>
<li>The test statistics are all related. The <em>t</em> statistic from the t-test is <strong>1.532</strong>, which is the same as the t-statistic from the linear regression. If you square that, you get <strong>2.347</strong>, the <em>F</em> statistic from the ANOVA.</li>
<li>The <code><a href="https://rdrr.io/r/stats/t.test.html">t.test()</a></code> output shows you the means for the two groups, Committed and Single. Just displaying the <code>fit</code> object itself or running <code>summary(fit)</code> shows you the coefficients for a linear model. Here, the model assumes the “baseline” RelationshipStatus level is <em>Committed</em>, and that the <em>intercept</em> in a regression model (e.g., <span class="math inline">\(\beta_{0}\)</span> in the model <span class="math inline">\(Y = \beta_{0} + \beta_{1}X\)</span>) is the mean of the baseline group. Being <em>Single</em> results in an increase in BMI of 0.3413. This is the <span class="math inline">\(\beta_{1}\)</span> coefficient in the model. You can easily change the ordering of the levels. See the help for <code><a href="https://rdrr.io/r/base/factor.html">?factor</a></code>, and check out the new <a href="http://forcats.tidyverse.org/"><strong>forcats</strong> package</a>, which provides tools <strong>for</strong> manipulating <strong>cat</strong>egorical variables.</li>
</ol>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-21_eba9f4f337cce2e48b5689aa8e3cd551">
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># P-value computed on a t-statistic with 3552 degrees of freedom</span></span>
<span><span class="co"># (multiply times 2 because t-test is assuming two-tailed)</span></span>
<span><span class="fl">2</span><span class="op">*</span><span class="op">(</span><span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/stats/TDist.html">pt</a></span><span class="op">(</span><span class="fl">1.532</span>, df<span class="op">=</span><span class="fl">3552</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1256115</code></pre>
</div>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># P-value computed on an F-test with 1 and 3552 degrees of freedom</span></span>
<span><span class="fl">1</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/stats/Fdist.html">pf</a></span><span class="op">(</span><span class="fl">2.347</span>, df1<span class="op">=</span><span class="fl">1</span>, df2<span class="op">=</span><span class="fl">3552</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1256134</code></pre>
</div>
</div>
<blockquote class="blockquote">
<p><i class="fa fa-exclamation-triangle" aria-hidden="true"></i> <em><strong>A note on dummy coding:</strong></em> If you have a <span class="math inline">\(k\)</span>-level factor, R creates <span class="math inline">\(k-1\)</span> dummy variables, or indicator variables, by default, using the alphabetically first level as baseline. For example, the levels of RelationshipStatus are “Committed” and “Single”. R creates a dummy variable called “RelationshipStatusSingle” that’s <strong>0</strong> if you’re committed, and <strong>1</strong> if you’re Single. The linear model is saying for every unit increase in RelationshipStatusSingle, i.e., going from committed to single, results in a 0.314-unit increase in BMI. You can change the ordering of the factors to change the interpretation of the model (e.g., treating Single as baseline and going from Single to Committed). We’ll do this in the next section.</p>
</blockquote>
</section><section id="anova" class="level3" data-number="8.3.4"><h3 data-number="8.3.4" class="anchored" data-anchor-id="anova">
<span class="header-section-number">8.3.4</span> ANOVA</h3>
<p>Recap: t-tests are for assessing the differences in means between <em>two</em> groups. A t-test is a specific case of ANOVA, which is a specific case of a linear model. Let’s run ANOVA, but this time looking for differences in means between more than two groups.</p>
<p>Let’s look at the relationship between smoking status (Never, Former, or Current), and BMI.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-22_5e78e6a5da9b053c14f503fc97226dbf">
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">BMI</span><span class="op">~</span><span class="va">SmokingStatus</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/anova.html">anova</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Analysis of Variance Table

Response: BMI
                Df Sum Sq Mean Sq F value   Pr(&gt;F)
SmokingStatus    2   1411  705.50  16.988 4.54e-08
Residuals     3553 147551   41.53                 </code></pre>
</div>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = BMI ~ SmokingStatus, data = nha)

Residuals:
    Min      1Q  Median      3Q     Max 
-12.565  -4.556  -1.056   3.315  51.744 

Coefficients:
                    Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)          27.3914     0.2446 111.975  &lt; 2e-16
SmokingStatusFormer   1.7736     0.3293   5.387 7.65e-08
SmokingStatusNever    1.4645     0.2838   5.161 2.60e-07

Residual standard error: 6.444 on 3553 degrees of freedom
  (151 observations deleted due to missingness)
Multiple R-squared:  0.009472,  Adjusted R-squared:  0.008915 
F-statistic: 16.99 on 2 and 3553 DF,  p-value: 4.54e-08</code></pre>
</div>
</div>
<p>The F-test on the ANOVA table tells us that there <em>is</em> a significant difference in means between current, former, and never smokers (p=<span class="math inline">\(4.54 \times 10^{-8}\)</span>). However, the linear model output might not have been what we wanted. Because the default handling of categorical variables is to treat the alphabetical first level as the baseline, “Current” smokers are treated as baseline, and this mean becomes the intercept, and the coefficients on “Former” and “Never” describe how those groups’ means differ from current smokers.</p>
<p>Back to dummy coding / indicator variables: SmokingStatus is “Current”, “Former”, and “Never.” By default, R will create <em>two</em> indicator variables here that in tandem will explain this variable.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-23_86435e2bb9b8bfaeccd7721844b188f7">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 27%">
<col style="width: 36%">
<col style="width: 35%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">Original SmokingStatus</th>
<th style="text-align: right;">Indicator: SmokingStatusFormer</th>
<th style="text-align: right;">Indicator: SmokingStatusNever</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><em>Current</em></td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;"><em>Former</em></td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><em>Never</em></td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>What if we wanted “Never” smokers to be the baseline, followed by Former, then Current? Have a look at <code><a href="https://rdrr.io/r/base/factor.html">?factor</a></code> to relevel the factor levels.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-24_116387c60f849a489e6ea6e1a1167aa7">
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Look at nha$SmokingStatus</span></span>
<span><span class="va">nha</span><span class="op">$</span><span class="va">SmokingStatus</span></span>
<span></span>
<span><span class="co"># What happens if we relevel it? Let's see what that looks like.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">nha</span><span class="op">$</span><span class="va">SmokingStatus</span>, ref<span class="op">=</span><span class="st">"Never"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># If we're happy with that, let's change the value of nha$SmokingStatus in place</span></span>
<span><span class="va">nha</span><span class="op">$</span><span class="va">SmokingStatus</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">nha</span><span class="op">$</span><span class="va">SmokingStatus</span>, ref<span class="op">=</span><span class="st">"Never"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or we could do this the dplyr way</span></span>
<span><span class="va">nha</span> <span class="op">&lt;-</span> <span class="va">nha</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>SmokingStatus<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">SmokingStatus</span>, ref<span class="op">=</span><span class="st">"Never"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-25_9ddb2359c32597851bf12988dae7b3b6">
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Re-fit the model</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">BMI</span><span class="op">~</span><span class="va">SmokingStatus</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Optionally, show the ANOVA table</span></span>
<span><span class="co"># anova(fit)</span></span>
<span></span>
<span><span class="co"># Print the full model statistics</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = BMI ~ SmokingStatus, data = nha)

Residuals:
    Min      1Q  Median      3Q     Max 
-12.565  -4.556  -1.056   3.315  51.744 

Coefficients:
                     Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)           28.8558     0.1438 200.601  &lt; 2e-16
SmokingStatusCurrent  -1.4645     0.2838  -5.161  2.6e-07
SmokingStatusFormer    0.3091     0.2632   1.175     0.24

Residual standard error: 6.444 on 3553 degrees of freedom
  (151 observations deleted due to missingness)
Multiple R-squared:  0.009472,  Adjusted R-squared:  0.008915 
F-statistic: 16.99 on 2 and 3553 DF,  p-value: 4.54e-08</code></pre>
</div>
</div>
<p>Notice that the p-value on the ANOVA/regression didn’t change, but the coefficients did. <em>Never</em> smokers are now treated as baseline. The intercept coefficient (28.856) is now the mean for <em>Never</em> smokers. The <code>SmokingStatusFormer</code> coefficient of .309 shows the apparent increase in BMI that former smokers have when compared to never smokers, but that difference is not significant (p=.24). The <code>SmokingStatusCurrent</code> coefficient of -1.464 shows that current smokers actually have a lower BMI than never smokers, and that this decrease is highly significant.</p>
<p>Finally, you can do the typical post-hoc ANOVA procedures on the fit object. For example, the <code><a href="https://rdrr.io/r/stats/TukeyHSD.html">TukeyHSD()</a></code> function will run <a href="https://en.wikipedia.org/wiki/Tukey%27s_range_test"><em>Tukey’s test</em></a> (also known as <em>Tukey’s range test</em>, the <em>Tukey method</em>, <em>Tukey’s honest significance test</em>, <em>Tukey’s HSD test</em> (honest significant difference), or the <em>Tukey-Kramer method</em>). Tukey’s test computes all pairwise mean difference calculation, comparing each group to each other group, identifying any difference between two groups that’s greater than the standard error, while controlling the type I error for all multiple comparisons. First run <code><a href="https://rdrr.io/r/stats/aov.html">aov()</a></code> (<strong>not</strong> <code><a href="https://rdrr.io/r/stats/anova.html">anova()</a></code>) on the fitted linear model object, then run <code><a href="https://rdrr.io/r/stats/TukeyHSD.html">TukeyHSD()</a></code> on the resulting analysis of variance fit.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-26_06b738cf3816b414ea43c1082194cacb">
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/TukeyHSD.html">TukeyHSD</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/aov.html">aov</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Tukey multiple comparisons of means
    95% family-wise confidence level

Fit: aov(formula = fit)

$SmokingStatus
                     diff        lwr        upr     p adj
Current-Never  -1.4644502 -2.1298249 -0.7990756 0.0000008
Former-Never    0.3091076 -0.3079639  0.9261790 0.4685044
Former-Current  1.7735578  1.0015525  2.5455631 0.0000002</code></pre>
</div>
</div>
<p>This shows that there isn’t much of a difference between former and never smokers, but that both of these differ significantly from current smokers, who have significantly lower BMI.</p>
<p>Finally, let’s visualize the differences in means between these groups. The <strong>NA</strong> category, which is omitted from the ANOVA, contains all the observations who have missing or non-recorded Smoking Status.</p>
<div class="cell" data-hash="stats_cache/html/smoking_boxplots_836133602465d5aa5eee332e47db7ba9">
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">nha</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">SmokingStatus</span>, <span class="va">BMI</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html">geom_boxplot</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_classic</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/smoking_boxplots-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="linear-regression" class="level3" data-number="8.3.5"><h3 data-number="8.3.5" class="anchored" data-anchor-id="linear-regression">
<span class="header-section-number">8.3.5</span> Linear regression</h3>
<p>Linear models are mathematical representations of the process that (<em>we think</em>) gave rise to our data. The model seeks to explain the relationship between a variable of interest, our <em>Y</em>, <em>outcome</em>, <em>response</em>, or <em>dependent</em> variable, and one or more <em>X</em>, <em>predictor</em>, or <em>independent</em> variables. Previously we talked about t-tests or ANOVA in the context of a simple linear regression model with only a single predictor variable, <span class="math inline">\(X\)</span>:</p>
<p><span class="math display">\[Y = \beta_{0} +  \beta_{1}X\]</span></p>
<p>But you can have multiple predictors in a linear model that are all additive, accounting for the effects of the others:</p>
<p><span class="math display">\[Y = \beta_{0} +  \beta_{1}X_{1} + \beta_{2}X_{2} + \epsilon\]</span></p>
<ul>
<li>
<span class="math inline">\(Y\)</span> is the response</li>
<li>
<span class="math inline">\(X_{1}\)</span> and <span class="math inline">\(X_{2}\)</span> are the predictors</li>
<li>
<span class="math inline">\(\beta_{0}\)</span> is the intercept, and <span class="math inline">\(\beta_{1}\)</span>, <span class="math inline">\(\beta_{2}\)</span> etc are <em>coefficients</em> that describe what 1-unit changes in <span class="math inline">\(X_{1}\)</span> and <span class="math inline">\(X_{2}\)</span> do to the outcome variable <span class="math inline">\(Y\)</span>.</li>
<li>
<span class="math inline">\(\epsilon\)</span> is random error. Our model will not perfectly predict <span class="math inline">\(Y\)</span>. It will be off by some random amount. We assume this amount is a random draw from a Normal distribution with mean 0 and standard deviation <span class="math inline">\(\sigma\)</span>.</li>
</ul>
<p><em>Building a linear model</em> means we propose a linear model and then estimate the coefficients and the variance of the error term. Above, this means estimating <span class="math inline">\(\beta_{0}, \beta_{1}, \beta_{2}\)</span> and <span class="math inline">\(\sigma\)</span>. This is what we do in R.</p>
<p>Let’s look at the relationship between height and weight.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-27_b04e793e3fa93e40f25855f63c08dc39">
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">Weight</span><span class="op">~</span><span class="va">Height</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = Weight ~ Height, data = nha)

Residuals:
    Min      1Q  Median      3Q     Max 
-40.339 -13.109  -2.658   9.309 127.972 

Coefficients:
             Estimate Std. Error t value Pr(&gt;|t|)
(Intercept) -73.70590    5.08110  -14.51   &lt;2e-16
Height        0.91996    0.03003   30.63   &lt;2e-16

Residual standard error: 18.57 on 3674 degrees of freedom
  (31 observations deleted due to missingness)
Multiple R-squared:  0.2034,    Adjusted R-squared:  0.2032 
F-statistic: 938.4 on 1 and 3674 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>The relationship is highly significant (P&lt;<span class="math inline">\(2.2 \times 10^{-16}\)</span>). The intercept term is not very useful most of the time. Here it shows us what the value of Weight would be when Height=0, which could never happen. The Height coefficient is meaningful – each one unit increase in height results in a 0.92 increase in the corresponding unit of weight. Let’s visualize that relationship:</p>
<div class="cell" data-hash="stats_cache/html/scatter_height_weight_lm_630a2db4b0eeb7453fb38f604c18e329">
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">nha</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Height</span>, y<span class="op">=</span><span class="va">Weight</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"lm"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/scatter_height_weight_lm-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>By default, this is only going to show the prediction over the range of the data. This is important! You never want to try to extrapolate response variables outside of the range of your predictor(s). For example, the linear model tells us that weight is -73.7kg when height is zero. We can extend the predicted model / regression line past the lowest value of the data down to height=0. The bands on the confidence interval tell us that the model is apparently confident within the regions defined by the gray boundary. But this is silly – we would never see a height of zero, and predicting past the range of the available training data is never a good idea.</p>
<div class="cell" data-hash="stats_cache/html/scatter_height_weight_extrapolate_c970cd32ae0d5c311058aceae7500fe1">
<div class="sourceCode" id="cb65"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="va">nha</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span>x<span class="op">=</span><span class="va">Height</span>, y<span class="op">=</span><span class="va">Weight</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html">geom_smooth</a></span><span class="op">(</span>method<span class="op">=</span><span class="st">"lm"</span>, fullrange<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html">xlim</a></span><span class="op">(</span><span class="fl">0</span>, <span class="cn">NA</span><span class="op">)</span> <span class="op">+</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html">ggtitle</a></span><span class="op">(</span><span class="st">"Friends don't let friends extrapolate."</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/scatter_height_weight_extrapolate-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section><section id="multiple-regression" class="level3" data-number="8.3.6"><h3 data-number="8.3.6" class="anchored" data-anchor-id="multiple-regression">
<span class="header-section-number">8.3.6</span> Multiple regression</h3>
<p>Finally, let’s do a multiple linear regression analysis, where we attempt to model the effect of multiple predictor variables at once on some outcome. First, let’s look at the effect of physical activity on testosterone levels. Let’s do this with a t-test and linear regression, showing that you get the same results.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-28_74b541f6a698619f47e4f100e722ecaa">
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">Testosterone</span><span class="op">~</span><span class="va">PhysActive</span>, data<span class="op">=</span><span class="va">nha</span>, var.equal<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Two Sample t-test

data:  Testosterone by PhysActive
t = -2.4298, df = 3436, p-value = 0.01516
alternative hypothesis: true difference in means between group No and group Yes is not equal to 0
95 percent confidence interval:
 -34.813866  -3.720171
sample estimates:
 mean in group No mean in group Yes 
         207.5645          226.8315 </code></pre>
</div>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">Testosterone</span><span class="op">~</span><span class="va">PhysActive</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = Testosterone ~ PhysActive, data = nha)

Residuals:
   Min     1Q Median     3Q    Max 
-224.5 -196.5 -115.9  167.0 1588.0 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)    207.565      5.873   35.34   &lt;2e-16
PhysActiveYes   19.267      7.929    2.43   0.0152

Residual standard error: 231.4 on 3436 degrees of freedom
  (269 observations deleted due to missingness)
Multiple R-squared:  0.001715,  Adjusted R-squared:  0.001425 
F-statistic: 5.904 on 1 and 3436 DF,  p-value: 0.01516</code></pre>
</div>
</div>
<p>In both cases, the p-value is significant (p=0.01516), and the result suggest that increased physical activity is associated with increased testosterone levels. Does increasing your physical activity increase your testosterone levels? Or is it the other way – will increased testosterone encourage more physical activity? Or is it none of the above – is the apparent relationship between physical activity and testosterone levels only apparent because both are correlated with yet a third, unaccounted for variable? Let’s throw Age into the model as well.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-29_6e4ff25e4c7f411d0b9bad4f22bf0688">
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">Testosterone</span><span class="op">~</span><span class="va">PhysActive</span><span class="op">+</span><span class="va">Age</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = Testosterone ~ PhysActive + Age, data = nha)

Residuals:
   Min     1Q Median     3Q    Max 
-238.6 -196.8 -112.3  167.4 1598.1 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)   247.8828    13.0853  18.944  &lt; 2e-16
PhysActiveYes  13.6740     8.0815   1.692 0.090735
Age            -0.8003     0.2322  -3.447 0.000574

Residual standard error: 231 on 3435 degrees of freedom
  (269 observations deleted due to missingness)
Multiple R-squared:  0.005156,  Adjusted R-squared:  0.004577 
F-statistic: 8.901 on 2 and 3435 DF,  p-value: 0.0001394</code></pre>
</div>
</div>
<p>This shows us that after accounting for age that the testosterone / physical activity link is no longer significant. Every 1-year increase in age results in a highly significant decrease in testosterone, and since increasing age is also likely associated with decreased physical activity, perhaps age is the confounder that makes this relationship apparent.</p>
<p>Adding other predictors can also swing things the other way. We know that men have much higher testosterone levels than females. Sex is probably the single best predictor of testosterone levels in our dataset. By not accounting for this effect, our unaccounted-for variation remains very high. By accounting for Gender, we now reduce the residual error in the model, and the physical activity effect once again becomes significant. Also notice that our model fits much better (higher R-squared), and is much more significant overall.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-30_24fc0e0123907212014c0099e333b50e">
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">Testosterone</span> <span class="op">~</span> <span class="va">PhysActive</span><span class="op">+</span><span class="va">Age</span><span class="op">+</span><span class="va">Gender</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = Testosterone ~ PhysActive + Age + Gender, data = nha)

Residuals:
    Min      1Q  Median      3Q     Max 
-397.91  -31.01   -4.42   20.50 1400.90 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)    46.6931     7.5729   6.166 7.83e-10
PhysActiveYes   9.2749     4.4617   2.079   0.0377
Age            -0.5904     0.1282  -4.605 4.28e-06
Gendermale    385.1989     4.3512  88.526  &lt; 2e-16

Residual standard error: 127.5 on 3434 degrees of freedom
  (269 observations deleted due to missingness)
Multiple R-squared:  0.6969,    Adjusted R-squared:  0.6966 
F-statistic:  2632 on 3 and 3434 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>We’ve only looked at the <a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/summary.lm.html"><code>summary()</code></a> and <a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/anova.lm.html"><code>anova()</code></a> functions for extracting information from an <a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/lm.html"><code>lm</code> class object</a>. There are several other accessor functions that can be used on a linear model object. Check out the help page for each one of these to learn more.</p>
<ul>
<li><a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/coef.html"><code>coefficients()</code></a></li>
<li><a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/predict.lm.html"><code>predict.lm()</code></a></li>
<li><a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/fitted.values.html"><code>fitted.values()</code></a></li>
<li><a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/residuals.html"><code>residuals()</code></a></li>
</ul>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 5
</div>
</div>
<div class="callout-body-container callout-body">
<p>Is the average BMI different in single people versus those in a committed relationship? Perform a t-test.</p>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 6
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>Work</code> variable is coded “Looking” (n=159), “NotWorking” (n=1317), and “Working” (n=2230).</p>
<ul>
<li>Fit a linear model of <code>Income</code> against <code>Work</code>. Assign this to an object called <code>fit</code>. What does the <code>fit</code> object tell you when you display it directly?</li>
<li>Run an <code><a href="https://rdrr.io/r/stats/anova.html">anova()</a></code> to get the ANOVA table. Is the model significant?</li>
<li>Run a Tukey test to get the pairwise contrasts. (Hint: <code><a href="https://rdrr.io/r/stats/TukeyHSD.html">TukeyHSD()</a></code> on <code><a href="https://rdrr.io/r/stats/aov.html">aov()</a></code> on the fit). What do you conclude?</li>
<li>Instead of thinking of this as ANOVA, think of it as a linear model. After you’ve thought about it, get some <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> statistics on the fit. Do these results jive with the ANOVA model?</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 7
</div>
</div>
<div class="callout-body-container callout-body">
<p>Examine the relationship between HDL cholesterol levels (<code>HDLChol</code>) and whether someone has diabetes or not (<code>Diabetes</code>).</p>
<ul>
<li>Is there a difference in means between diabetics and nondiabetics? Perform a t-test <em>without</em> a Welch correction (that is, assuming equal variances – see <code><a href="https://rdrr.io/r/stats/t.test.html">?t.test</a></code> for help).</li>
<li>Do the same analysis in a linear modeling framework.</li>
<li>Does the relationship hold when adjusting for <code>Weight</code>?</li>
<li>What about when adjusting for <code>Weight</code>, <code>Age</code>, <code>Gender</code>, <code>PhysActive</code> (whether someone participates in moderate or vigorous-intensity sports, fitness or recreational activities, coded as yes/no). What is the effect of each of these explanatory variables?</li>
</ul>
</div>
</div>
</section></section><section id="discrete-variables" class="level2" data-number="8.4"><h2 data-number="8.4" class="anchored" data-anchor-id="discrete-variables">
<span class="header-section-number">8.4</span> Discrete variables</h2>
<p>Until now we’ve only discussed analyzing <em>continuous</em> outcomes / dependent variables. We’ve tested for differences in means between two groups with t-tests, differences among means between <em>n</em> groups with ANOVA, and more general relationships using linear regression. In all of these cases, the dependent variable, i.e., the outcome, or <span class="math inline">\(Y\)</span> variable, was <em>continuous</em>, and usually normally distributed. What if our outcome variable is <em>discrete</em>, e.g., “Yes/No”, “Mutant/WT”, “Case/Control”, etc.? Here we use a different set of procedures for assessing significant associations.</p>
<section id="contingency-tables" class="level3" data-number="8.4.1"><h3 data-number="8.4.1" class="anchored" data-anchor-id="contingency-tables">
<span class="header-section-number">8.4.1</span> Contingency tables</h3>
<p>The <a href="http://stat.ethz.ch/R-manual/R-patched/library/stats/html/xtabs.html"><code>xtabs()</code></a> function is useful for creating contingency tables from categorical variables. Let’s create a gender by diabetes status contingency table, and assign it to an object called <strong><code>xt</code></strong>. After making the assignment, type the name of the object to view it.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-32_12e1375b57befb4f4228d0b858640298">
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">xt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/xtabs.html">xtabs</a></span><span class="op">(</span><span class="op">~</span><span class="va">Gender</span><span class="op">+</span><span class="va">Diabetes</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span>
<span><span class="va">xt</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Diabetes
Gender     No  Yes
  female 1692  164
  male   1653  198</code></pre>
</div>
</div>
<p>There are two useful functions, <code><a href="https://rdrr.io/r/stats/addmargins.html">addmargins()</a></code> and <code><a href="https://rdrr.io/r/base/proportions.html">prop.table()</a></code> that add more information or manipulate how the data is displayed. By default, <code><a href="https://rdrr.io/r/base/proportions.html">prop.table()</a></code> will divide the number of observations in each cell by the total. But you may want to specify <em>which margin</em> you want to get proportions over. Let’s do this for the first (row) margin.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-33_bbec0a7dd1e34ec9b62da9414f5a5140">
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Add marginal totals</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/addmargins.html">addmargins</a></span><span class="op">(</span><span class="va">xt</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Diabetes
Gender     No  Yes  Sum
  female 1692  164 1856
  male   1653  198 1851
  Sum    3345  362 3707</code></pre>
</div>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Get the proportional table</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span><span class="op">(</span><span class="va">xt</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Diabetes
Gender           No        Yes
  female 0.45643377 0.04424063
  male   0.44591314 0.05341246</code></pre>
</div>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># That wasn't really what we wanted. </span></span>
<span><span class="co"># Do this over the first (row) margin only.</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span><span class="op">(</span><span class="va">xt</span>, margin<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Diabetes
Gender           No        Yes
  female 0.91163793 0.08836207
  male   0.89303079 0.10696921</code></pre>
</div>
</div>
<p>Looks like men have slightly higher rates of diabetes than women. But is this significant?</p>
<p>The chi-square test is used to assess the independence of these two factors. That is, if the null hypothesis that gender and diabetes are independent is true, the we would expect a proportionally equal number of diabetics across each sex. Males seem to be at slightly higher risk than females, but the difference is just short of statistically significant.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-34_7b04d7737a087c69f0877eaabca8f8f7">
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/chisq.test.html">chisq.test</a></span><span class="op">(</span><span class="va">xt</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Pearson's Chi-squared test with Yates' continuity correction

data:  xt
X-squared = 3.4332, df = 1, p-value = 0.0639</code></pre>
</div>
</div>
<p>An alternative to the chi-square test is <a href="https://en.wikipedia.org/wiki/Fisher%27s_exact_test">Fisher’s exact test</a>. Rather than relying on a critical value from a theoretical chi-square distribution, Fisher’s exact test calculates the <em>exact</em> probability of observing the contingency table as is. It’s especially useful when there are very small <em>n</em>’s in one or more of the contingency table cells. Both the chi-square and Fisher’s exact test give us p-values of approximately 0.06.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-35_140ad9a9a12ca6c21f581a825de73af6">
<div class="sourceCode" id="cb84"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/fisher.test.html">fisher.test</a></span><span class="op">(</span><span class="va">xt</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Fisher's Exact Test for Count Data

data:  xt
p-value = 0.05992
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
 0.9883143 1.5466373
sample estimates:
odds ratio 
  1.235728 </code></pre>
</div>
</div>
<p>There’s a useful plot for visualizing contingency table data called a <em>mosaic</em> plot. Call the <code><a href="https://rdrr.io/r/graphics/mosaicplot.html">mosaicplot()</a></code> function on the contingency table object. Note this is a built-in plot, <em>not</em> a ggplot2-style plot.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-36_6e5b4d07a8fbf47839434400c1617293">
<div class="sourceCode" id="cb86"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/mosaicplot.html">mosaicplot</a></span><span class="op">(</span><span class="va">xt</span>, main<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/unnamed-chunk-36-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Let’s create a different contingency table, this time looking at the relationship between race and whether the person had health insurance. Display the table with marginal totals.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-37_4a597743cbbb7ff195b5864f253b849c">
<div class="sourceCode" id="cb87"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">xt</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/xtabs.html">xtabs</a></span><span class="op">(</span><span class="op">~</span><span class="va">Race</span><span class="op">+</span><span class="va">Insured</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/addmargins.html">addmargins</a></span><span class="op">(</span><span class="va">xt</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Insured
Race         No  Yes  Sum
  Asian      46  169  215
  Black      86  330  416
  Hispanic   89  151  240
  Mexican   147  141  288
  Other      33   65   98
  White     307 2141 2448
  Sum       708 2997 3705</code></pre>
</div>
</div>
<p>Let’s do the same thing as above, this time showing the proportion of people in each race category having health insurance.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-38_72ce5a232459f7e4ea487830fd187679">
<div class="sourceCode" id="cb89"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/proportions.html">prop.table</a></span><span class="op">(</span><span class="va">xt</span>, margin<span class="op">=</span><span class="fl">1</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Insured
Race              No       Yes
  Asian    0.2139535 0.7860465
  Black    0.2067308 0.7932692
  Hispanic 0.3708333 0.6291667
  Mexican  0.5104167 0.4895833
  Other    0.3367347 0.6632653
  White    0.1254085 0.8745915</code></pre>
</div>
</div>
<p>Now, let’s run a chi-square test for independence.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-39_c1465391068181b4436f86bc6187060d">
<div class="sourceCode" id="cb91"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/chisq.test.html">chisq.test</a></span><span class="op">(</span><span class="va">xt</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Pearson's Chi-squared test

data:  xt
X-squared = 323.3, df = 5, p-value &lt; 2.2e-16</code></pre>
</div>
</div>
<p>The result is <em>highly</em> significant. In fact, so significant, that the display rounds off the p-value to something like <span class="math inline">\(&lt;2.2 \times 10^{-16}\)</span>. If you look at the help for <a href="https://stat.ethz.ch/R-manual/R-patched/library/stats/html/chisq.test.html"><code>?chisq.test</code></a> you’ll see that displaying the test only shows you summary information, but other components can be accessed. For example, we can easily get the actual p-value, or the expected counts under the null hypothesis of independence.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-40_619aacdb9ef3a777c50d963c2a2e23f8">
<div class="sourceCode" id="cb93"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/chisq.test.html">chisq.test</a></span><span class="op">(</span><span class="va">xt</span><span class="op">)</span><span class="op">$</span><span class="va">p.value</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 9.754238e-68</code></pre>
</div>
<div class="sourceCode" id="cb95"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/chisq.test.html">chisq.test</a></span><span class="op">(</span><span class="va">xt</span><span class="op">)</span><span class="op">$</span><span class="va">expected</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          Insured
Race              No        Yes
  Asian     41.08502  173.91498
  Black     79.49474  336.50526
  Hispanic  45.86235  194.13765
  Mexican   55.03482  232.96518
  Other     18.72713   79.27287
  White    467.79595 1980.20405</code></pre>
</div>
</div>
<p>We can also make a mosaic plot similar to above:</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-41_ae80eb5c24f5ed198c236094a3a1730f">
<div class="sourceCode" id="cb97"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/mosaicplot.html">mosaicplot</a></span><span class="op">(</span><span class="va">xt</span>, main<span class="op">=</span><span class="cn">NA</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid" width="768"></p>
</div>
</div>
</section><section id="logistic-regression" class="level3" data-number="8.4.2"><h3 data-number="8.4.2" class="anchored" data-anchor-id="logistic-regression">
<span class="header-section-number">8.4.2</span> Logistic regression</h3>
<p><a href="slides/r-stats.html#(8)"><em>(See slides)</em></a></p>
<p>What if we wanted to model the discrete outcome, e.g., whether someone is insured, against several other variables, similar to how we did with multiple linear regression? We can’t use linear regression because the outcome isn’t continuous – it’s binary, either <em>Yes</em> or <em>No</em>. For this we’ll use <em>logistic regression</em> to model the <em>log odds</em> of binary response. That is, instead of modeling the outcome variable, <span class="math inline">\(Y\)</span>, directly against the inputs, we’ll model the <em>log odds</em> of the outcome variable.</p>
<p>If <span class="math inline">\(p\)</span> is the probability that the individual is insured, then <span class="math inline">\(\frac{p}{1-p}\)</span> is the <a href="https://en.wikipedia.org/wiki/Odds"><em>odds</em></a> that person is insured. Then it follows that the linear model is expressed as:</p>
<p><span class="math display">\[log(\frac{p}{1-p}) = \beta_0 + \beta_1 x_1 + \cdots + \beta_k x_k\]</span></p>
<p>Where <span class="math inline">\(\beta_0\)</span> is the intercept, <span class="math inline">\(\beta_1\)</span> is the increase in the odds of the outcome for every unit increase in <span class="math inline">\(x_1\)</span>, and so on.</p>
<p>Logistic regression is a type of <em>generalized linear model</em> (GLM). We fit GLM models in R using the <code><a href="https://rdrr.io/r/stats/glm.html">glm()</a></code> function. It works like the <code><a href="https://rdrr.io/r/stats/lm.html">lm()</a></code> function except we specify which GLM to fit using the <code>family</code> argument. Logistic regression requires <code>family=binomial</code>.</p>
<p>The typical use looks like this:</p>
<div class="sourceCode" id="cb98"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">mod</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">y</span> <span class="op">~</span> <span class="va">x</span>, data<span class="op">=</span><span class="va">yourdata</span>, family<span class="op">=</span><span class="st">'binomial'</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">mod</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Before we fit a logistic regression model let’s <em>relevel</em> the Race variable so that “White” is the baseline. We saw above that people who identify as “White” have the highest rates of being insured. When we run the logistic regression, we’ll get a separate coefficient (effect) for each level of the factor variable(s) in the model, telling you the increased odds that that level has, <em>as compared to the baseline group</em>.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-42_943fd9dbfbb7ca654185b92e62ec0a81">
<div class="sourceCode" id="cb99"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#Look at Race. The default ordering is alphabetical</span></span>
<span><span class="va">nha</span><span class="op">$</span><span class="va">Race</span></span>
<span></span>
<span><span class="co"># Let's relevel that where the group with the highest rate of insurance is "baseline"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">nha</span><span class="op">$</span><span class="va">Race</span>, ref<span class="op">=</span><span class="st">"White"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># If we're happy with that result, permanently change it</span></span>
<span><span class="va">nha</span><span class="op">$</span><span class="va">Race</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">nha</span><span class="op">$</span><span class="va">Race</span>, ref<span class="op">=</span><span class="st">"White"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Or do it the dplyr way</span></span>
<span><span class="va">nha</span> <span class="op">&lt;-</span> <span class="va">nha</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>Race<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/relevel.html">relevel</a></span><span class="op">(</span><span class="va">Race</span>, ref<span class="op">=</span><span class="st">"White"</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s fit a logistic regression model assessing how the odds of being insured change with different levels of race.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-43_32de3172b4dce79c3d200d320cf2f679">
<div class="sourceCode" id="cb100"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">Insured</span><span class="op">~</span><span class="va">Race</span>, data<span class="op">=</span><span class="va">nha</span>, family<span class="op">=</span><span class="st">"binomial"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = Insured ~ Race, family = "binomial", data = nha)

Coefficients:
             Estimate Std. Error z value Pr(&gt;|z|)
(Intercept)   1.94218    0.06103  31.825  &lt; 2e-16
RaceAsian    -0.64092    0.17715  -3.618 0.000297
RaceBlack    -0.59744    0.13558  -4.406 1.05e-05
RaceHispanic -1.41354    0.14691  -9.622  &lt; 2e-16
RaceMexican  -1.98385    0.13274 -14.946  &lt; 2e-16
RaceOther    -1.26430    0.22229  -5.688 1.29e-08

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 3614.6  on 3704  degrees of freedom
Residual deviance: 3336.6  on 3699  degrees of freedom
  (2 observations deleted due to missingness)
AIC: 3348.6

Number of Fisher Scoring iterations: 4</code></pre>
</div>
</div>
<p>The <code>Estimate</code> column shows the log of the odds ratio – how the log odds of having health insurance changes at each level of race compared to White. The P-value for each coefficient is on the far right. This shows that <em>every</em> other race has <em>significantly less</em> rates of health insurance coverage. But, as in our multiple linear regression analysis above, are there other important variables that we’re leaving out that could alter our conclusions? Lets add a few more variables into the model to see if something else can explain the apparent Race-Insured association. Let’s add a few things likely to be involved (Age and Income), and something that’s probably irrelevant (hours slept at night).</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-44_28b22392fe3a949b294c7c05dde2ae08">
<div class="sourceCode" id="cb102"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">Insured</span> <span class="op">~</span> <span class="va">Age</span><span class="op">+</span><span class="va">Income</span><span class="op">+</span><span class="va">SleepHrsNight</span><span class="op">+</span><span class="va">Race</span>, data<span class="op">=</span><span class="va">nha</span>, family<span class="op">=</span><span class="st">"binomial"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = Insured ~ Age + Income + SleepHrsNight + Race, 
    family = "binomial", data = nha)

Coefficients:
                Estimate Std. Error z value Pr(&gt;|z|)
(Intercept)   -3.501e-01  2.919e-01  -1.199    0.230
Age            3.371e-02  2.949e-03  11.431  &lt; 2e-16
Income         1.534e-05  1.537e-06   9.982  &lt; 2e-16
SleepHrsNight -1.763e-02  3.517e-02  -0.501    0.616
RaceAsian     -4.550e-01  2.031e-01  -2.241    0.025
RaceBlack     -2.387e-01  1.536e-01  -1.554    0.120
RaceHispanic  -1.010e+00  1.635e-01  -6.175 6.61e-10
RaceMexican   -1.404e+00  1.483e-01  -9.468  &lt; 2e-16
RaceOther     -9.888e-01  2.422e-01  -4.082 4.46e-05

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 3284.3  on 3395  degrees of freedom
Residual deviance: 2815.0  on 3387  degrees of freedom
  (311 observations deleted due to missingness)
AIC: 2833

Number of Fisher Scoring iterations: 5</code></pre>
</div>
</div>
<p>A few things become apparent:</p>
<ol type="1">
<li>Age and income are both highly associated with whether someone is insured. Both of these variables are highly significant (<span class="math inline">\(P&lt;2.2 \times 10^{-16}\)</span>), and the coefficient (the <code>Estimate</code> column) is positive, meaning that for each unit increase in one of these variables, the odds of being insured increases by the corresponding amount.</li>
<li>Hours slept per night is not meaningful at all.</li>
<li>After accounting for age and income, several of the race-specific differences are no longer statistically significant, but others remain so.</li>
<li>The absolute value of the test statistic (column called <code>z value</code>) can roughly be taken as an estimate of the “importance” of that variable to the overall model. So, age and income are the most important influences in this model; self-identifying as Hispanic or Mexican are also very highly important, hours slept per night isn’t important at all, and the other race categories fall somewhere in between.</li>
</ol>
<p>There is <em>much</em> more to go into with logistic regression. This chapter only scratches the surface. Missing from this chapter are things like regression diagnostics, model comparison approaches, penalization, interpretation of model coefficients, fitting interaction effects, and much more. Alan Agresti’s <a href="https://www.amazon.com/Categorical-Data-Analysis-Alan-Agresti/dp/0470463635/ref=sr_1_1?ie=UTF8&amp;qid=1473180895&amp;sr=8-1&amp;keywords=categorical+data+analysis&amp;tag=gettgenedone-20"><em>Categorical Data Analysis</em></a> has long been considered the definitive text on this topic. I also recommend Agresti’s <a href="https://www.amazon.com/Introduction-Categorical-Data-Analysis/dp/0471226181/ref=sr_1_3?ie=UTF8&amp;qid=1473180895&amp;sr=8-3&amp;keywords=categorical+data+analysis&amp;tag=gettgenedone-20"><em>Introduction to Categorical Data Analysis</em></a> (a.k.a. “Agresti lite”) for a gentler introduction.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 8
</div>
</div>
<div class="callout-body-container callout-body">
<p>What’s the relationship between diabetes and participating in rigorous physical activity or sports?</p>
<ul>
<li>Create a contingency table with Diabetes status in rows and physical activity status in columns.</li>
<li>Display that table with margins.</li>
<li>Show the proportions of diabetics and nondiabetics, separately, who are physically active or not.</li>
<li>Is this relationship significant?</li>
<li>Create a mosaic plot to visualize the relationship.</li>
</ul>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 9
</div>
</div>
<div class="callout-body-container callout-body">
<p>Model the same association in a logistic regression framework to assess the risk of diabetes using physical activity as a predictor.</p>
<!-- - First, make Diabetes a factor variable if you haven't already (`nha$Diabetes <- factor(nha$Diabetes)`). -->
<ul>
<li>Fit a model with just physical activity as a predictor, and display a model summary.</li>
<li>Add gender to the model, and show a summary.</li>
<li>Continue adding weight and age to the model. What happens to the gender association?</li>
<li>Continue and add income to the model. What happens to the original association with physical activity?</li>
</ul>
</div>
</div>
</section></section><section id="power-sample-size" class="level2" data-number="8.5"><h2 data-number="8.5" class="anchored" data-anchor-id="power-sample-size">
<span class="header-section-number">8.5</span> Power &amp; sample size</h2>
<!-- [_(See slides)_](slides/r-stats.html#(10)) -->
<p>This is a necessarily short introduction to the concept of power and sample size calculations. <a href="https://en.wikipedia.org/wiki/Statistical_power">Statistical power</a>, also sometimes called sensitivity, is defined as the probability that your test correctly rejects the null hypothesis when the alternative hypothesis is true. That is, if there really is an effect (difference in means, association between categorical variables, etc.), how likely are you to be able to <em>detect</em> that effect at a given statistical significance level, given certain assumptions. Generally there are a few moving pieces, and if you know all but one of them, you can calculate what that last one is.</p>
<ol type="1">
<li>Power: How likely are you to detect the effect? (Usually like to see 80% or greater).</li>
<li>N: What is the sample size you have (or require)?</li>
<li>Effect size: How big is the difference in means, odds ratio, etc?</li>
</ol>
<p>If we know we want 80% power to detect a certain magnitude of difference between groups, we can calculate our required sample size. Or, if we know we can only collect 5 samples, we can calculate how likely we are to detect a particular effect. Or, we can work to solve the last one - if we want 80% power and we have 5 samples, what’s the smallest effect we can hope to detect?</p>
<p>All of these questions require certain assumptions about the data and the testing procedure. Which kind of test is being performed? What’s the true effect size (often unknown, or estimated from preliminary data), what’s the standard deviation of samples that will be collected (often unknown, or estimated from preliminary data), what’s the level of statistical significance needed (traditionally p&lt;0.05, but must consider multiple testing corrections).</p>
<section id="t-test-powern" class="level3" data-number="8.5.1"><h3 data-number="8.5.1" class="anchored" data-anchor-id="t-test-powern">
<span class="header-section-number">8.5.1</span> T-test power/N</h3>
<p>The <code><a href="https://rdrr.io/r/stats/power.t.test.html">power.t.test()</a></code> empirically estimates power or sample size of a t-test for differences in means. If we have 20 samples in each of two groups (e.g., control versus treatment), and the standard deviation for whatever we’re measuring is <strong>2.3</strong>, and we’re expecting a true difference in means between the groups of <strong>2</strong>, what’s the power to detect this effect?</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-46_aa6f15e58ad1e15fe7119b6ab2b0a2fd">
<div class="sourceCode" id="cb104"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/power.t.test.html">power.t.test</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">20</span>, delta<span class="op">=</span><span class="fl">2</span>, sd<span class="op">=</span><span class="fl">2.3</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Two-sample t test power calculation 

              n = 20
          delta = 2
             sd = 2.3
      sig.level = 0.05
          power = 0.7641668
    alternative = two.sided

NOTE: n is number in *each* group</code></pre>
</div>
</div>
<p>What’s the sample size we’d need to detect a difference of 0.8 given a standard deviation of 1.5, assuming we want 80% power?</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-47_197c0512d37ba695b55d5fd668e2ae94">
<div class="sourceCode" id="cb106"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/power.t.test.html">power.t.test</a></span><span class="op">(</span>power<span class="op">=</span><span class="fl">.80</span>, delta<span class="op">=</span><span class="fl">.8</span>, sd<span class="op">=</span><span class="fl">1.5</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Two-sample t test power calculation 

              n = 56.16413
          delta = 0.8
             sd = 1.5
      sig.level = 0.05
          power = 0.8
    alternative = two.sided

NOTE: n is number in *each* group</code></pre>
</div>
</div>
</section><section id="proportions-powern" class="level3" data-number="8.5.2"><h3 data-number="8.5.2" class="anchored" data-anchor-id="proportions-powern">
<span class="header-section-number">8.5.2</span> Proportions power/N</h3>
<p>What about a two-sample proportion test (e.g., chi-square test)? If we have two groups (control and treatment), and we’re measuring some outcome (e.g., infected yes/no), and we know that the proportion of infected controls is 80% but 20% in treated, what’s the power to detect this effect in 5 samples per group?</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-48_a6f7af61b8a2564753cc41705c305447">
<div class="sourceCode" id="cb108"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/power.prop.test.html">power.prop.test</a></span><span class="op">(</span>n<span class="op">=</span><span class="fl">5</span>, p1<span class="op">=</span><span class="fl">0.8</span>, p2<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Two-sample comparison of proportions power calculation 

              n = 5
             p1 = 0.8
             p2 = 0.2
      sig.level = 0.05
          power = 0.4688159
    alternative = two.sided

NOTE: n is number in *each* group</code></pre>
</div>
</div>
<p>How many samples would we need for 90% power?</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-49_9a2cdfa8bce2ed38e4ea4fdde6d2c5dc">
<div class="sourceCode" id="cb110"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/power.prop.test.html">power.prop.test</a></span><span class="op">(</span>power<span class="op">=</span><span class="fl">0.9</span>, p1<span class="op">=</span><span class="fl">0.8</span>, p2<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
     Two-sample comparison of proportions power calculation 

              n = 12.37701
             p1 = 0.8
             p2 = 0.2
      sig.level = 0.05
          power = 0.9
    alternative = two.sided

NOTE: n is number in *each* group</code></pre>
</div>
</div>
<p>Also check out the <a href="https://cran.r-project.org/web/packages/pwr/"><strong>pwr</strong> package</a> which has power calculation functions for other statistical tests.</p>
<table class="table">
<thead><tr class="header">
<th>Function</th>
<th>Power calculations for</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>pwr.2p.test()</code></td>
<td>Two proportions (equal n)</td>
</tr>
<tr class="even">
<td><code>pwr.2p2n.test()</code></td>
<td>Two proportions (unequal n)</td>
</tr>
<tr class="odd">
<td><code>pwr.anova.test()</code></td>
<td>Balanced one way ANOVA</td>
</tr>
<tr class="even">
<td><code>pwr.chisq.test()</code></td>
<td>Chi-square test</td>
</tr>
<tr class="odd">
<td><code>pwr.f2.test()</code></td>
<td>General linear model</td>
</tr>
<tr class="even">
<td><code>pwr.p.test()</code></td>
<td>Proportion (one sample)</td>
</tr>
<tr class="odd">
<td><code>pwr.r.test()</code></td>
<td>Correlation</td>
</tr>
<tr class="even">
<td><code>pwr.t.test()</code></td>
<td>T-tests (one sample, 2 sample, paired)</td>
</tr>
<tr class="odd">
<td><code>pwr.t2n.test()</code></td>
<td>T-test (two samples with unequal n)</td>
</tr>
</tbody>
</table>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 10
</div>
</div>
<div class="callout-body-container callout-body">
<p>You’re doing a gene expression experiment. What’s your power to detect a 2-fold change in a gene with a standard deviation of 0.7, given 3 samples? (Note - fold change is usually given on the <span class="math inline">\(log_2\)</span> scale, so a 2-fold change would be a <code>delta</code> of 1. That is, if the fold change is 2x, then <span class="math inline">\(log_2(2)=1\)</span>, and you should use 1 in the calculation, not 2).</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-50_9b5a841b9b9f2191e0e9026b87f579a2">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.2709095</code></pre>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 11
</div>
</div>
<div class="callout-body-container callout-body">
<p>How many samples would you need to have 80% power to detect this effect?</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-51_4735b2640d3c829df0f4d5c66a258c5c">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 8.764711</code></pre>
</div>
</div>
</div>
</div>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 12
</div>
</div>
<div class="callout-body-container callout-body">
<p>You’re doing a population genome-wide association study (GWAS) looking at the effect of a SNP on disease X. Disease X has a baseline prevalence of 5% in the population, but you suspect the SNP might increase the risk of disease X by 10% (this is typical for SNP effects on common, complex diseases). Calculate the number of samples do you need to have 80% power to detect this effect, given that you want a genome-wide statistical significance of <span class="math inline">\(p&lt;5\times10^{-8}\)</span> to account for multiple testing.<a href="#fn1" class="footnote-ref" id="fnref1" role="doc-noteref"><sup>1</sup></a> (Hint, you can expressed <span class="math inline">\(5\times10^{-8}\)</span> in R using <code>5e-8</code> instead of <code>.00000005</code>).</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-52_57aaedbb5d38bd781884de95748b4842">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 157589.5</code></pre>
</div>
</div>
</div>
</div>
</section></section><section id="tidying-models" class="level2" data-number="8.6"><h2 data-number="8.6" class="anchored" data-anchor-id="tidying-models">
<span class="header-section-number">8.6</span> Tidying models</h2>
<!-- [_(See slides)_](slides/r-stats.html#(11)) -->
<p>We spent a lot of time in previous chapters on <em>tidy <strong>data</strong></em>, where each column is a variable and each row is an observation. Tidy data is easy to filter observations based on values in a column (e.g., we could get just adult males with <code>filter(nha, Gender=="male" &amp; Age&gt;=18)</code>, and easy to select particular variables/features of interest by their column name.</p>
<p>Even when we start with tidy <em>data</em>, we don’t end up with tidy <em>models</em>. The output from tests like <code>t.test</code> or <code>lm</code> are not data.frames, and it’s difficult to get the information out of the model object that we want. The <a href="https://github.com/dgrtwo/broom"><strong>broom</strong> package</a> bridges this gap.</p>
<p><img src="img/broom.jpg" class="img-fluid"></p>
<p>Depending on the type of model object you’re using, broom provides three methods that do different kinds of tidying:</p>
<ol type="1">
<li>
<code>tidy</code>: constructs a data frame that summarizes the model’s statistical findings like coefficients and p-values.</li>
<li>
<code>augment</code>: add columns to the original data that was modeled, like predictions and residuals.</li>
<li>
<code>glance</code>: construct a concise <em>one-row</em> summary of the model with information like <span class="math inline">\(R^2\)</span> that are computed once for the entire model.</li>
</ol>
<p>Let’s go back to our linear model example.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-53_322b2051367bfcae59e90023cccd2d1d">
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Try modeling Testosterone against Physical Activity, Age, and Gender.</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">Testosterone</span><span class="op">~</span><span class="va">PhysActive</span><span class="op">+</span><span class="va">Age</span><span class="op">+</span><span class="va">Gender</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># See what that model looks like:</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = Testosterone ~ PhysActive + Age + Gender, data = nha)

Residuals:
    Min      1Q  Median      3Q     Max 
-397.91  -31.01   -4.42   20.50 1400.90 

Coefficients:
              Estimate Std. Error t value Pr(&gt;|t|)
(Intercept)    46.6931     7.5729   6.166 7.83e-10
PhysActiveYes   9.2749     4.4617   2.079   0.0377
Age            -0.5904     0.1282  -4.605 4.28e-06
Gendermale    385.1989     4.3512  88.526  &lt; 2e-16

Residual standard error: 127.5 on 3434 degrees of freedom
  (269 observations deleted due to missingness)
Multiple R-squared:  0.6969,    Adjusted R-squared:  0.6966 
F-statistic:  2632 on 3 and 3434 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>What if we wanted to pull out the coefficient for Age, or the P-value for PhysActive? It gets pretty gross. We first have to <code>coef(summary(lmfit))</code> to get a matrix of coefficients, the terms are still stored in row names, and the column names are inconsistent with other packages (e.g.&nbsp;<code>Pr(&gt;|t|)</code> compared to <code>p.value</code>). Yuck!</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-54_6b741e297b466cc8993881385efeeb4c">
<div class="sourceCode" id="cb117"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="st">"Age"</span>, <span class="st">"Estimate"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] -0.5903616</code></pre>
</div>
<div class="sourceCode" id="cb119"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="st">"PhysActiveYes"</span>, <span class="st">"Pr(&gt;|t|)"</span><span class="op">]</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.03771185</code></pre>
</div>
</div>
<p>Instead, you can use the <code>tidy</code> function, from the broom package, on the fit:</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-55_b1c5fefc8aabf33ffcc66dbbd1f25b0a">
<div class="sourceCode" id="cb121"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Install the package if you don't have it</span></span>
<span><span class="co"># install.packages("broom")</span></span>
<span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://broom.tidymodels.org/">broom</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 5
  term          estimate std.error statistic  p.value
  &lt;chr&gt;            &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)     46.7       7.57       6.17 7.83e-10
2 PhysActiveYes    9.27      4.46       2.08 3.77e- 2
3 Age             -0.590     0.128     -4.60 4.28e- 6
4 Gendermale     385.        4.35      88.5  0       </code></pre>
</div>
</div>
<p>This gives you a data.frame with all your model results. The row names have been moved into a column called <code>term</code>, and the column names are simple and consistent (and can be accessed using <code>$</code>). These can be manipulated with dplyr just like any other data frame.</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-56_062461a3884655949e848b3173a30003">
<div class="sourceCode" id="cb123"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">term</span><span class="op">!=</span><span class="st">"(Intercept)"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">term</span>, <span class="va">p.value</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  term             p.value
  &lt;chr&gt;              &lt;dbl&gt;
1 Gendermale    0         
2 Age           0.00000428
3 PhysActiveYes 0.0377    </code></pre>
</div>
</div>
<p>Instead of viewing the coefficients, you might be interested in the fitted values and residuals for each of the original points in the regression. For this, use <code>augment</code>, which augments the original data with information from the model. New columns begins with a <code>.</code> (to avoid overwriting any of the original columns).</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-57_b941458ec906d1e85cfc03b8a59649f7">
<div class="sourceCode" id="cb125"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Augment the original data</span></span>
<span><span class="co"># IF you get a warning about deprecated... purrr..., ignore. It's a bug that'll be fixed soon.</span></span>
<span><span class="fu"><a href="https://generics.r-lib.org/reference/augment.html">augment</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 11
  .rownames Testosterone PhysActive   Age Gender .fitted .resid     .hat .sigma
  &lt;chr&gt;            &lt;dbl&gt; &lt;fct&gt;      &lt;dbl&gt; &lt;fct&gt;    &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt;
1 1                 47.5 No            43 female    21.3  26.2  0.000989   128.
2 2                643.  No            80 male     385.  258.   0.00185    127.
3 3                643.  No            80 male     385.  258.   0.00185    127.
4 4                 21.1 Yes           34 female    35.9 -14.8  0.000928   128.
5 5                563.  No            80 male     385.  178.   0.00185    128.
6 6                402.  No            35 male     411.   -9.45 0.00117    128.
# ℹ 2 more variables: .cooksd &lt;dbl&gt;, .std.resid &lt;dbl&gt;</code></pre>
</div>
<div class="sourceCode" id="cb127"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># Plot residuals vs fitted values for males, </span></span>
<span><span class="co"># colored by Physical Activity, size scaled by age</span></span>
<span><span class="fu"><a href="https://generics.r-lib.org/reference/augment.html">augment</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">Gender</span><span class="op">==</span><span class="st">"male"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">.fitted</span>, <span class="va">.resid</span>, col<span class="op">=</span><span class="va">PhysActive</span>, size<span class="op">=</span><span class="va">Age</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html">geom_point</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Finally, several summary statistics are computed for the entire regression, such as <span class="math inline">\(R^2\)</span> and the F-statistic. These can be accessed with <code>glance</code>:</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-58_fd9400694b0f565b6c708084402b2c65">
<div class="sourceCode" id="cb128"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://generics.r-lib.org/reference/glance.html">glance</a></span><span class="op">(</span><span class="va">fit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 12
  r.squared adj.r.squared sigma statistic p.value    df  logLik    AIC    BIC
      &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;   &lt;dbl&gt; &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
1     0.697         0.697  128.     2632.       0     3 -21545. 43100. 43130.
# ℹ 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</code></pre>
</div>
</div>
<p>The <strong>broom</strong> functions work on a pipe, so you can <code>|&gt;</code> your model directly to any of the functions like <code><a href="https://generics.r-lib.org/reference/tidy.html">tidy()</a></code>. Let’s tidy up our t-test:</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-59_f663dc267ce800c0ba1713431a541907">
<div class="sourceCode" id="cb130"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">AlcoholYear</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Welch Two Sample t-test

data:  AlcoholYear by RelationshipStatus
t = 5.4315, df = 2674.8, p-value = 6.09e-08
alternative hypothesis: true difference in means between group Committed and group Single is not equal to 0
95 percent confidence interval:
 13.05949 27.81603
sample estimates:
mean in group Committed    mean in group Single 
               83.93416                63.49640 </code></pre>
</div>
<div class="sourceCode" id="cb132"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/t.test.html">t.test</a></span><span class="op">(</span><span class="va">AlcoholYear</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 10
  estimate estimate1 estimate2 statistic    p.value parameter conf.low conf.high
     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;      &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt;
1     20.4      83.9      63.5      5.43    6.09e-8     2675.     13.1      27.8
# ℹ 2 more variables: method &lt;chr&gt;, alternative &lt;chr&gt;</code></pre>
</div>
</div>
<p>…and our Mann-Whitney <em>U</em> test / Wilcoxon rank-sum test:</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-60_fbcbe3c364712263f3e72dc0f69e7f9e">
<div class="sourceCode" id="cb134"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html">wilcox.test</a></span><span class="op">(</span><span class="va">AlcoholYear</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Wilcoxon rank sum test with continuity correction

data:  AlcoholYear by RelationshipStatus
W = 1067954, p-value = 0.0001659
alternative hypothesis: true location shift is not equal to 0</code></pre>
</div>
<div class="sourceCode" id="cb136"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/wilcox.test.html">wilcox.test</a></span><span class="op">(</span><span class="va">AlcoholYear</span><span class="op">~</span><span class="va">RelationshipStatus</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 4
  statistic  p.value method                                          alternative
      &lt;dbl&gt;    &lt;dbl&gt; &lt;chr&gt;                                           &lt;chr&gt;      
1  1067954. 0.000166 Wilcoxon rank sum test with continuity correct… two.sided  </code></pre>
</div>
</div>
<p>…and our Fisher’s exact test on the cross-tabulated data:</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-61_de6d78a244f36914c52a0ed3d8a1dda1">
<div class="sourceCode" id="cb138"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/xtabs.html">xtabs</a></span><span class="op">(</span><span class="op">~</span><span class="va">Gender</span><span class="op">+</span><span class="va">Diabetes</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/fisher.test.html">fisher.test</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
    Fisher's Exact Test for Count Data

data:  xtabs(~Gender + Diabetes, data = nha)
p-value = 0.05992
alternative hypothesis: true odds ratio is not equal to 1
95 percent confidence interval:
 0.9883143 1.5466373
sample estimates:
odds ratio 
  1.235728 </code></pre>
</div>
<div class="sourceCode" id="cb140"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/stats/xtabs.html">xtabs</a></span><span class="op">(</span><span class="op">~</span><span class="va">Gender</span><span class="op">+</span><span class="va">Diabetes</span>, data<span class="op">=</span><span class="va">nha</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/stats/fisher.test.html">fisher.test</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 6
  estimate p.value conf.low conf.high method                         alternative
     &lt;dbl&gt;   &lt;dbl&gt;    &lt;dbl&gt;     &lt;dbl&gt; &lt;chr&gt;                          &lt;chr&gt;      
1     1.24  0.0599    0.988      1.55 Fisher's Exact Test for Count… two.sided  </code></pre>
</div>
</div>
<p>…and finally, a logistic regression model:</p>
<div class="cell" data-hash="stats_cache/html/unnamed-chunk-62_9fdf265446ac5b7de9acc97e25e2ba13">
<div class="sourceCode" id="cb142"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># fit the model and summarize it the usual way</span></span>
<span><span class="va">glmfit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/glm.html">glm</a></span><span class="op">(</span><span class="va">Insured</span><span class="op">~</span><span class="va">Race</span>, data<span class="op">=</span><span class="va">nha</span>, family<span class="op">=</span><span class="va">binomial</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">glmfit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
glm(formula = Insured ~ Race, family = binomial, data = nha)

Coefficients:
             Estimate Std. Error z value Pr(&gt;|z|)
(Intercept)   1.94218    0.06103  31.825  &lt; 2e-16
RaceAsian    -0.64092    0.17715  -3.618 0.000297
RaceBlack    -0.59744    0.13558  -4.406 1.05e-05
RaceHispanic -1.41354    0.14691  -9.622  &lt; 2e-16
RaceMexican  -1.98385    0.13274 -14.946  &lt; 2e-16
RaceOther    -1.26430    0.22229  -5.688 1.29e-08

(Dispersion parameter for binomial family taken to be 1)

    Null deviance: 3614.6  on 3704  degrees of freedom
Residual deviance: 3336.6  on 3699  degrees of freedom
  (2 observations deleted due to missingness)
AIC: 3348.6

Number of Fisher Scoring iterations: 4</code></pre>
</div>
<div class="sourceCode" id="cb144"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># tidy it up!</span></span>
<span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="va">glmfit</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 5
  term         estimate std.error statistic   p.value
  &lt;chr&gt;           &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)     1.94     0.0610     31.8  2.96e-222
2 RaceAsian      -0.641    0.177      -3.62 2.97e-  4
3 RaceBlack      -0.597    0.136      -4.41 1.05e-  5
4 RaceHispanic   -1.41     0.147      -9.62 6.47e- 22
5 RaceMexican    -1.98     0.133     -14.9  1.66e- 50
6 RaceOther      -1.26     0.222      -5.69 1.29e-  8</code></pre>
</div>
<div class="sourceCode" id="cb146"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># do whatever you want now</span></span>
<span><span class="fu"><a href="https://generics.r-lib.org/reference/tidy.html">tidy</a></span><span class="op">(</span><span class="va">glmfit</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">term</span><span class="op">!=</span><span class="st">"(Intercept)"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>logp<span class="op">=</span><span class="op">-</span><span class="fl">1</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html">log10</a></span><span class="op">(</span><span class="va">p.value</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">term</span>, <span class="va">logp</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html">geom_bar</a></span><span class="op">(</span>stat<span class="op">=</span><span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/coord_flip.html">coord_flip</a></span><span class="op">(</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="stats_files/figure-html/unnamed-chunk-62-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>Check out some of the other <a href="https://cran.r-project.org/web/packages/broom/index.html"><strong>broom</strong> vignettes on CRAN</a>, and also check out the <a href="http://bioconductor.org/packages/release/bioc/html/biobroom.html"><strong>biobroom</strong> package</a> on bioconductor for turning bioconductor objects and analytical results into tidy data frames.</p>
<!-- ## Homework -->
<!-- ### Stress tests & muscular dystrophy genetics -->
<!-- **[Try this homework assignment](r-stats-homework.html).**  -->
<!-- There are two parts to this assignment, each using data curated and hosted by the [Vanderbilt Department of Biostatistics](http://biostat.mc.vanderbilt.edu/wiki/Main/DataSets). The first part looks at the effects of different doses of a drug used during echocardiograms called dobutamine. The second uses data collected to examine several blood serum markers believed to be associated with genetics for a specific kind of muscular dystrophy (DMD).  -->
</section><section id="additional-topics-recommended-reading" class="level2" data-number="8.7"><h2 data-number="8.7" class="anchored" data-anchor-id="additional-topics-recommended-reading">
<span class="header-section-number">8.7</span> Additional topics &amp; recommended reading</h2>
<section id="batch-effects" class="level3" data-number="8.7.1"><h3 data-number="8.7.1" class="anchored" data-anchor-id="batch-effects">
<span class="header-section-number">8.7.1</span> 1. Batch effects</h3>
<p><em>Batch effects</em> are sources of technical variation introduced during an experiment, such as processing with different reagents, handling by a different technician, sequencing on a different flow cell, or processing samples in groups on different days. If these <em>batch effects</em> are strongly confounded with the study variable of interest, they can call into question the validity of your results, and in some cases, render collected data completely useless. The papers below discuss batch effects and how they can be mitigated.</p>
<ol type="1">
<li>
<strong>Chapter 5</strong> of Scherer, Andreas. <em>Batch effects and noise in microarray experiments: sources and solutions.</em> Vol. 868. John Wiley &amp; Sons, 2009.
<ul>
<li>Chapter 5 only: <a href="http://onlinelibrary.wiley.com/doi/10.1002/9780470685983.ch5/pdf" class="uri">http://onlinelibrary.wiley.com/doi/10.1002/9780470685983.ch5/pdf</a>.</li>
<li>Entire book: <a href="https://faculty.mu.edu.sa/public/uploads/1382673974.78419780470741382.pdf" class="uri">https://faculty.mu.edu.sa/public/uploads/1382673974.78419780470741382.pdf</a>.</li>
</ul>
</li>
<li>Leek, Jeffrey T., et al.&nbsp;“Tackling the widespread and critical impact of batch effects in high-throughput data.” <em>Nature Reviews Genetics</em> 11.10 (2010): 733-739. Available at <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3880143/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3880143/</a>.</li>
</ol></section><section id="whats-my-n" class="level3" data-number="8.7.2"><h3 data-number="8.7.2" class="anchored" data-anchor-id="whats-my-n">
<span class="header-section-number">8.7.2</span> 2. What’s my <em>n</em>?</h3>
<p>“What’s my <em>n</em>” isn’t always a straightforward question to answer, especially when it comes to cell culture expriments. The post and article below go into some of these details.</p>
<ol type="1">
<li>Statistics for Experimental Biologists: “What is ‘n’ in cell culture experiments?” Available at <a href="http://labstats.net/articles/cell_culture_n.html" class="uri">http://labstats.net/articles/cell_culture_n.html</a>.</li>
<li>Vaux, David L., Fiona Fidler, and Geoff Cumming. “Replicates and repeats—what is the difference and is it significant?.” <em>EMBO reports</em> 13.4 (2012): 291-296. Available at <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3321166/" class="uri">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3321166/</a>.</li>
</ol></section><section id="technical-versus-biological-replicates" class="level3" data-number="8.7.3"><h3 data-number="8.7.3" class="anchored" data-anchor-id="technical-versus-biological-replicates">
<span class="header-section-number">8.7.3</span> 3. Technical versus biological replicates</h3>
<p>Technical replicates involve taking multiple measurements on the same sample. Biological replicates are different samples each with separate measurements/assays. While technical replicates can help calibrate the precision of an instrument or assay, biological replicates are necessary for statistical analysis to make inferences about a condition or treatment. Read the paper and note below for more information on technical vs biological replication.</p>
<ol type="1">
<li>Blainey, Paul, Martin Krzywinski, and Naomi Altman. “Points of significance: replication.” <em>Nature methods</em> 11.9 (2014): 879-880. Available at <a href="http://rdcu.be/yguA" class="uri">http://rdcu.be/yguA</a>.</li>
<li>Illumina Technical Note: “The Power of Replicates.” Available at <a href="https://www.illumina.com/Documents/products/technotes/technote_power_replicates.pdf" class="uri">https://www.illumina.com/Documents/products/technotes/technote_power_replicates.pdf</a>.</li>
</ol>


</section></section><section id="footnotes" class="footnotes footnotes-end-of-document" role="doc-endnotes"><hr>
<ol>
<li id="fn1"><p><a href="https://www.quora.com/Why-is-P-value-5x10-8-chosen-as-a-threshold-to-reach-genome-wide-significance" class="uri">https://www.quora.com/Why-is-P-value-5x10-8-chosen-as-a-threshold-to-reach-genome-wide-significance</a><a href="#fnref1" class="footnote-back" role="doc-backlink">↩︎</a></p></li>
</ol></section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./rmarkdown.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Reproducible Reporting with RMarkdown</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./survival.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Survival Analysis</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">© 2019 Stephen D. Turner, Ph.D.</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://x.com/strnr">
      <i class="bi bi-twitter" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/stephenturner">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/turnersd/">
      <i class="bi bi-linkedin" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://stephenturner.us/">
      <i class="bi bi-house-fill" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://blog.stephenturner.us/">
      <i class="bi bi-rss" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>