<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.313">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Stephen D. Turner, Ph.D.">
<title>Biological Data Science with R - 2&nbsp; Tibbles</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./dplyr.html" rel="next">
<link href="./basics.html" rel="prev">
<link href="./hex.png" rel="icon" type="image/png">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>
</head>
<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./basics.html">Core Curriculum</a></li><li class="breadcrumb-item"><a href="./tibbles.html"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Tibbles</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto"><div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Biological Data Science with R</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/stephenturner/bdsr" title="Source Code" class="quarto-navigation-tool px-1" aria-label="Source Code"><i class="bi bi-github"></i></a>
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
<li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Biological-Data-Science-with-R.pdf">
              <i class="bi bi-bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Biological-Data-Science-with-R.epub">
              <i class="bi bi-bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
      </ul>
</div>
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Core Curriculum</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Basics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tibbles.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Tibbles</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dplyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation with dplyr</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tidyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Tidy Data and Advanced Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ggplot2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Visualization with ggplot2</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tidyeda.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Refresher: Tidy Exploratory Data Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./rmarkdown.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Reproducible Reporting with RMarkdown</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Electives</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./stats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Essential statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./survival.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Survival Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./predmodeling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Predictive Analytics: Predicting and Forecasting Influenza</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./textmining.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Text Mining and NLP</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./rnaseq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Count-Based Differential Expression Analysis of RNA-seq Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ggtree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Visualizing and Annotating Phylogenetic Trees</span></span></a>
  </div>
</li>
      </ul>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Setup</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Further Resources</span></span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">Table of contents</h2>
   
  <ul>
<li><a href="#our-data" id="toc-our-data" class="nav-link active" data-scroll-target="#our-data"><span class="header-section-number">2.1</span> Our data</a></li>
  <li>
<a href="#reading-in-data" id="toc-reading-in-data" class="nav-link" data-scroll-target="#reading-in-data"><span class="header-section-number">2.2</span> Reading in data</a>
  <ul class="collapse">
<li><a href="#dplyr-and-readr" id="toc-dplyr-and-readr" class="nav-link" data-scroll-target="#dplyr-and-readr"><span class="header-section-number">2.2.1</span> dplyr and readr</a></li>
  <li><a href="#read_csv" id="toc-read_csv" class="nav-link" data-scroll-target="#read_csv"><span class="header-section-number">2.2.2</span> <code>read_csv()</code></a></li>
  </ul>
</li>
  <li>
<a href="#inspecting-data.frame-objects" id="toc-inspecting-data.frame-objects" class="nav-link" data-scroll-target="#inspecting-data.frame-objects"><span class="header-section-number">2.3</span> Inspecting data.frame objects</a>
  <ul class="collapse">
<li><a href="#built-in-functions" id="toc-built-in-functions" class="nav-link" data-scroll-target="#built-in-functions"><span class="header-section-number">2.3.1</span> Built-in functions</a></li>
  <li><a href="#other-packages" id="toc-other-packages" class="nav-link" data-scroll-target="#other-packages"><span class="header-section-number">2.3.2</span> Other packages</a></li>
  </ul>
</li>
  <li><a href="#accessing-variables-subsetting-data-frames" id="toc-accessing-variables-subsetting-data-frames" class="nav-link" data-scroll-target="#accessing-variables-subsetting-data-frames"><span class="header-section-number">2.4</span> Accessing variables &amp; subsetting data frames</a></li>
  <li><a href="#bonus-preview-to-advanced-manipulation" id="toc-bonus-preview-to-advanced-manipulation" class="nav-link" data-scroll-target="#bonus-preview-to-advanced-manipulation"><span class="header-section-number">2.5</span> BONUS: Preview to advanced manipulation</a></li>
  </ul></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title"><span id="sec-tibbles" class="quarto-section-identifier"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Tibbles</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header><p>There are <em>lots</em> of different basic data structures in R. If you take any kind of longer introduction to R you’ll probably learn about arrays, lists, matrices, etc. Let’s skip straight to the data structure you’ll probably use most – the <strong>data frame</strong>. We use data frames to store heterogeneous tabular data in R: tabular, meaning that individuals or observations are typically represented in rows, while variables or features are represented as columns; heterogeneous, meaning that columns/features/variables can be different classes (on variable, e.g.&nbsp;age, can be numeric, while another, e.g., cause of death, can be text).</p>
<p><strong>This chapter assumes a basic familiarity with R (see <a href="basics.html">Chapter&nbsp;<span>1</span></a>).</strong></p>
<p><strong>Recommended reading:</strong> Review the <a href="http://r4ds.had.co.nz/tibbles.html#introduction-4"><em>Introduction</em> (10.1)</a> and <a href="http://r4ds.had.co.nz/tibbles.html#tibbles-vs.data.frame"><em>Tibbles vs.&nbsp;data.frame</em> (10.3)</a> sections of the <a href="http://r4ds.had.co.nz/tibbles.html"><strong><em>R for Data Science</em> book</strong></a>. We will initially be using the <code>read_*</code> functions from the <a href="http://readr.tidyverse.org/"><strong>readr</strong> package</a>. These functions load data into a <em>tibble</em> instead of R’s traditional data.frame. Tibbles are data frames, but they tweak some older behaviors to make life a little easier. These sections explain the few key small differences between traditional data.frames and tibbles.</p>
<section id="our-data" class="level2" data-number="2.1"><h2 data-number="2.1" class="anchored" data-anchor-id="our-data">
<span class="header-section-number">2.1</span> Our data</h2>
<!-- There are some built-in data frames that ship with R that you'll often see people referencing in help forums or other places to demonstrate some functionality. The `mtcars` data frame has data extracted from the 1974 _Motor Trend_ magazine giving statistics about design and performance for 32 different vehicles. You can just type `mtcars` to look at the whole dataset. -->
<p>The data we’re going to look at is cleaned up version of a gene expression dataset from <a href="http://www.ncbi.nlm.nih.gov/pubmed/17959824">Brauer et al.&nbsp;Coordination of Growth Rate, Cell Cycle, Stress Response, and Metabolic Activity in Yeast (2008) <em>Mol Biol Cell</em> 19:352-367</a>. This data is from a gene expression microarray, and in this paper the authors are examining the relationship between growth rate and gene expression in yeast cultures limited by one of six different nutrients (glucose, leucine, ammonium, sulfate, phosphate, uracil). If you give yeast a rich media loaded with nutrients except restrict the supply of a <em>single</em> nutrient, you can control the growth rate to any rate you choose. By starving yeast of specific nutrients you can find genes that:</p>
<ol type="1">
<li>
<strong>Raise or lower their expression in response to growth rate</strong>. Growth-rate dependent expression patterns can tell us a lot about cell cycle control, and how the cell responds to stress. The authors found that expression of &gt;25% of all yeast genes is linearly correlated with growth rate, independent of the limiting nutrient. They also found that the subset of negatively growth-correlated genes is enriched for peroxisomal functions, and positively correlated genes mainly encode ribosomal functions.</li>
<li>
<strong>Respond differently when different nutrients are being limited</strong>. If you see particular genes that respond very differently when a nutrient is sharply restricted, these genes might be involved in the transport or metabolism of that specific nutrient.</li>
</ol>
<p>You can download the cleaned up version of the data <a href="data/data.zip">here</a>. The file is called <a href="data/brauer2007_tidy.csv"><strong>brauer2007_tidy.csv</strong></a>. Later on we’ll actually start with the original raw data (minimally processed) and manipulate it so that we can make it more amenable for analysis.</p>
</section><section id="reading-in-data" class="level2" data-number="2.2"><h2 data-number="2.2" class="anchored" data-anchor-id="reading-in-data">
<span class="header-section-number">2.2</span> Reading in data</h2>
<section id="dplyr-and-readr" class="level3" data-number="2.2.1"><h3 data-number="2.2.1" class="anchored" data-anchor-id="dplyr-and-readr">
<span class="header-section-number">2.2.1</span> dplyr and readr</h3>
<p>There are some built-in functions for reading in data in text files. These functions are <em>read-dot-something</em> – for example, <code><a href="https://rdrr.io/r/utils/read.table.html">read.csv()</a></code> reads in comma-delimited text data; <code><a href="https://rdrr.io/r/utils/read.table.html">read.delim()</a></code> reads in tab-delimited text, etc. We’re going to read in data a little bit differently here using the <a href="http://readr.tidyverse.org/">readr</a> package. When you load the readr package, you’ll have access to very similar looking functions, named <em>read-underscore-something</em> – e.g., <code><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv()</a></code>. You have to have the readr package installed to access these functions. Compared to the base functions, they’re <em>much</em> faster, they’re good at guessing the types of data in the columns, they don’t do some of the other silly things that the base functions do. We’re going to use another package later on called <a href="https://cran.r-project.org/web/packages/dplyr/index.html">dplyr</a>, and if you have the dplyr package loaded as well, and you read in the data with readr, the data will display nicely.</p>
<p>First let’s load those packages.</p>
<div class="cell" data-hash="tibbles_cache/html/loadpkgs_58520724d40697cb1b0851857d0e7980">
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://readr.tidyverse.org">readr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org">dplyr</a></span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If you see a warning that looks like this: <code>Error in library(packageName) : there is no package called 'packageName'</code>, then you don’t have the package installed correctly. See the setup chapter (<a href="setup.html">Appendix&nbsp;<span>A</span></a>).</p>
</section><section id="read_csv" class="level3" data-number="2.2.2"><h3 data-number="2.2.2" class="anchored" data-anchor-id="read_csv">
<span class="header-section-number">2.2.2</span> <code>read_csv()</code>
</h3>
<p>Now, let’s actually load the data. You can get help for the import function with <code><a href="https://readr.tidyverse.org/reference/read_delim.html">?read_csv</a></code>. When we load data we assign it to a variable just like any other, and we can choose a name for that data. Since we’re going to be referring to this data a lot, let’s give it a short easy name to type. I’m going to call it <code>ydat</code>. Once we’ve loaded it we can type the name of the object itself (<code>ydat</code>) to see it printed to the screen.</p>
<div class="cell" data-hash="tibbles_cache/html/loaddata_b0bff63e9b721957dfaaed1ca43aeaf6">
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ydat</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html">read_csv</a></span><span class="op">(</span>file<span class="op">=</span><span class="st">"data/brauer2007_tidy.csv"</span><span class="op">)</span></span>
<span><span class="va">ydat</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 198,430 × 7
   symbol systematic_name nutrient  rate expression bp                     mf   
   &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;    &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;                  &lt;chr&gt;
 1 SFB2   YNL049C         Glucose   0.05      -0.24 ER to Golgi transport  mole…
 2 &lt;NA&gt;   YNL095C         Glucose   0.05       0.28 biological process un… mole…
 3 QRI7   YDL104C         Glucose   0.05      -0.02 proteolysis and pepti… meta…
 4 CFT2   YLR115W         Glucose   0.05      -0.33 mRNA polyadenylylatio… RNA …
 5 SSO2   YMR183C         Glucose   0.05       0.05 vesicle fusion*        t-SN…
 6 PSP2   YML017W         Glucose   0.05      -0.69 biological process un… mole…
 7 RIB2   YOL066C         Glucose   0.05      -0.55 riboflavin biosynthes… pseu…
 8 VMA13  YPR036W         Glucose   0.05      -0.75 vacuolar acidification hydr…
 9 EDC3   YEL015W         Glucose   0.05      -0.24 deadenylylation-indep… mole…
10 VPS5   YOR069W         Glucose   0.05      -0.16 protein retention in … prot…
# ℹ 198,420 more rows</code></pre>
</div>
</div>
<p>Take a look at that output. The nice thing about loading dplyr and reading in data with readr is that data frames are displayed in a much more friendly way. This dataset has nearly 200,000 rows and 7 columns. When you import data this way and try to display the object in the console, instead of trying to display all 200,000 rows, you’ll only see about 10 by default. Also, if you have so many columns that the data would wrap off the edge of your screen, those columns will not be displayed, but you’ll see at the bottom of the output which, if any, columns were hidden from view. If you want to see the whole dataset, there are two ways to do this. First, you can click on the name of the data.frame in the <strong>Environment</strong> panel in RStudio. Or you could use the <code><a href="https://rdrr.io/r/utils/View.html">View()</a></code> function (<em>with a capital V</em>).</p>
<div class="cell" data-hash="tibbles_cache/html/view_4b6e4ec884f4682eb9e6c1704432176f">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/View.html">View</a></span><span class="op">(</span><span class="va">ydat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section><section id="inspecting-data.frame-objects" class="level2" data-number="2.3"><h2 data-number="2.3" class="anchored" data-anchor-id="inspecting-data.frame-objects">
<span class="header-section-number">2.3</span> Inspecting data.frame objects</h2>
<section id="built-in-functions" class="level3" data-number="2.3.1"><h3 data-number="2.3.1" class="anchored" data-anchor-id="built-in-functions">
<span class="header-section-number">2.3.1</span> Built-in functions</h3>
<p>There are several built-in functions that are useful for working with data frames.</p>
<ul>
<li>Content:
<ul>
<li>
<code><a href="https://rdrr.io/r/utils/head.html">head()</a></code>: shows the first few rows</li>
<li>
<code><a href="https://rdrr.io/r/utils/head.html">tail()</a></code>: shows the last few rows</li>
</ul>
</li>
<li>Size:
<ul>
<li>
<code><a href="https://rdrr.io/r/base/dim.html">dim()</a></code>: returns a 2-element vector with the number of rows in the first element, and the number of columns as the second element (the dimensions of the object)</li>
<li>
<code><a href="https://rdrr.io/r/base/nrow.html">nrow()</a></code>: returns the number of rows</li>
<li>
<code><a href="https://rdrr.io/r/base/nrow.html">ncol()</a></code>: returns the number of columns</li>
</ul>
</li>
<li>Summary:
<ul>
<li>
<code><a href="https://rdrr.io/r/base/colnames.html">colnames()</a></code> (or just <code><a href="https://rdrr.io/r/base/names.html">names()</a></code>): returns the column names</li>
<li>
<code><a href="https://rdrr.io/r/utils/str.html">str()</a></code>: structure of the object and information about the class, length and content of each column</li>
<li>
<code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code>: works differently depending on what kind of object you pass to it. Passing a data frame to the <code><a href="https://rdrr.io/r/base/summary.html">summary()</a></code> function prints out useful summary statistics about numeric column (min, max, median, mean, etc.)</li>
</ul>
</li>
</ul>
<div class="cell" data-hash="tibbles_cache/html/tibble_functions_253a6688fc2ee8d1ecba985589a29c42">
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">head</a></span><span class="op">(</span><span class="va">ydat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
  symbol systematic_name nutrient  rate expression bp                      mf   
  &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;    &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;                   &lt;chr&gt;
1 SFB2   YNL049C         Glucose   0.05      -0.24 ER to Golgi transport   mole…
2 &lt;NA&gt;   YNL095C         Glucose   0.05       0.28 biological process unk… mole…
3 QRI7   YDL104C         Glucose   0.05      -0.02 proteolysis and peptid… meta…
4 CFT2   YLR115W         Glucose   0.05      -0.33 mRNA polyadenylylation* RNA …
5 SSO2   YMR183C         Glucose   0.05       0.05 vesicle fusion*         t-SN…
6 PSP2   YML017W         Glucose   0.05      -0.69 biological process unk… mole…</code></pre>
</div>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html">tail</a></span><span class="op">(</span><span class="va">ydat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 7
  symbol systematic_name nutrient  rate expression bp                      mf   
  &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;    &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;                   &lt;chr&gt;
1 DOA1   YKL213C         Uracil     0.3       0.14 ubiquitin-dependent pr… mole…
2 KRE1   YNL322C         Uracil     0.3       0.28 cell wall organization… stru…
3 MTL1   YGR023W         Uracil     0.3       0.27 cell wall organization… mole…
4 KRE9   YJL174W         Uracil     0.3       0.43 cell wall organization… mole…
5 UTH1   YKR042W         Uracil     0.3       0.19 mitochondrion organiza… mole…
6 &lt;NA&gt;   YOL111C         Uracil     0.3       0.04 biological process unk… mole…</code></pre>
</div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dim.html">dim</a></span><span class="op">(</span><span class="va">ydat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 198430      7</code></pre>
</div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">ydat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "symbol"          "systematic_name" "nutrient"        "rate"           
[5] "expression"      "bp"              "mf"             </code></pre>
</div>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html">str</a></span><span class="op">(</span><span class="va">ydat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>spc_tbl_ [198,430 × 7] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ symbol         : chr [1:198430] "SFB2" NA "QRI7" "CFT2" ...
 $ systematic_name: chr [1:198430] "YNL049C" "YNL095C" "YDL104C" "YLR115W" ...
 $ nutrient       : chr [1:198430] "Glucose" "Glucose" "Glucose" "Glucose" ...
 $ rate           : num [1:198430] 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 0.05 ...
 $ expression     : num [1:198430] -0.24 0.28 -0.02 -0.33 0.05 -0.69 -0.55 -0.75 -0.24 -0.16 ...
 $ bp             : chr [1:198430] "ER to Golgi transport" "biological process unknown" "proteolysis and peptidolysis" "mRNA polyadenylylation*" ...
 $ mf             : chr [1:198430] "molecular function unknown" "molecular function unknown" "metalloendopeptidase activity" "RNA binding" ...
 - attr(*, "spec")=
  .. cols(
  ..   symbol = col_character(),
  ..   systematic_name = col_character(),
  ..   nutrient = col_character(),
  ..   rate = col_double(),
  ..   expression = col_double(),
  ..   bp = col_character(),
  ..   mf = col_character()
  .. )
 - attr(*, "problems")=&lt;externalptr&gt; </code></pre>
</div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="va">ydat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    symbol          systematic_name      nutrient              rate      
 Length:198430      Length:198430      Length:198430      Min.   :0.050  
 Class :character   Class :character   Class :character   1st Qu.:0.100  
 Mode  :character   Mode  :character   Mode  :character   Median :0.200  
                                                          Mean   :0.175  
                                                          3rd Qu.:0.250  
                                                          Max.   :0.300  
   expression         bp                 mf           
 Min.   :-6.50   Length:198430      Length:198430     
 1st Qu.:-0.29   Class :character   Class :character  
 Median : 0.00   Mode  :character   Mode  :character  
 Mean   : 0.00                                        
 3rd Qu.: 0.29                                        
 Max.   : 6.64                                        </code></pre>
</div>
</div>
</section><section id="other-packages" class="level3" data-number="2.3.2"><h3 data-number="2.3.2" class="anchored" data-anchor-id="other-packages">
<span class="header-section-number">2.3.2</span> Other packages</h3>
<p>The <code><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse()</a></code> function is available once you load the <strong>dplyr</strong> library, and it’s like <code><a href="https://rdrr.io/r/utils/str.html">str()</a></code> but its display is a little bit better.</p>
<div class="cell" data-hash="tibbles_cache/html/unnamed-chunk-2_8322ca831ae0fa05b19e98204acd9c19">
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="fu"><a href="https://pillar.r-lib.org/reference/glimpse.html">glimpse</a></span><span class="op">(</span><span class="va">ydat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 198,430
Columns: 7
$ symbol          &lt;chr&gt; "SFB2", NA, "QRI7", "CFT2", "SSO2", "PSP2", "RIB2", "V…
$ systematic_name &lt;chr&gt; "YNL049C", "YNL095C", "YDL104C", "YLR115W", "YMR183C",…
$ nutrient        &lt;chr&gt; "Glucose", "Glucose", "Glucose", "Glucose", "Glucose",…
$ rate            &lt;dbl&gt; 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, …
$ expression      &lt;dbl&gt; -0.24, 0.28, -0.02, -0.33, 0.05, -0.69, -0.55, -0.75, …
$ bp              &lt;chr&gt; "ER to Golgi transport", "biological process unknown",…
$ mf              &lt;chr&gt; "molecular function unknown", "molecular function unkn…</code></pre>
</div>
</div>
<p>The <strong>skimr</strong> package has a nice function, skim, that provides summary statistics the user can skim quickly to understand your data. You can install it with <code>install.packages("skimr")</code> if you don’t have it already.</p>
<div class="cell" data-hash="tibbles_cache/html/unnamed-chunk-3_da0f57c9d64600fae5946ec0259a2fd5">
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://docs.ropensci.org/skimr/">skimr</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://docs.ropensci.org/skimr/reference/skim.html">skim</a></span><span class="op">(</span><span class="va">ydat</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<caption>Data summary</caption>
<tbody>
<tr class="odd">
<td style="text-align: left;">Name</td>
<td style="text-align: left;">ydat</td>
</tr>
<tr class="even">
<td style="text-align: left;">Number of rows</td>
<td style="text-align: left;">198430</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Number of columns</td>
<td style="text-align: left;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">_______________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Column type frequency:</td>
<td style="text-align: left;"></td>
</tr>
<tr class="even">
<td style="text-align: left;">character</td>
<td style="text-align: left;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">numeric</td>
<td style="text-align: left;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">________________________</td>
<td style="text-align: left;"></td>
</tr>
<tr class="odd">
<td style="text-align: left;">Group variables</td>
<td style="text-align: left;">None</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: character</strong></p>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 21%">
<col style="width: 13%">
<col style="width: 18%">
<col style="width: 5%">
<col style="width: 5%">
<col style="width: 8%">
<col style="width: 12%">
<col style="width: 14%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">min</th>
<th style="text-align: right;">max</th>
<th style="text-align: right;">empty</th>
<th style="text-align: right;">n_unique</th>
<th style="text-align: right;">whitespace</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">symbol</td>
<td style="text-align: right;">47250</td>
<td style="text-align: right;">0.76</td>
<td style="text-align: right;">2</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">4210</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">systematic_name</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1.00</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">5536</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">nutrient</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1.00</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">9</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="even">
<td style="text-align: left;">bp</td>
<td style="text-align: right;">7663</td>
<td style="text-align: right;">0.96</td>
<td style="text-align: right;">7</td>
<td style="text-align: right;">82</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">880</td>
<td style="text-align: right;">0</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mf</td>
<td style="text-align: right;">7663</td>
<td style="text-align: right;">0.96</td>
<td style="text-align: right;">11</td>
<td style="text-align: right;">125</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1085</td>
<td style="text-align: right;">0</td>
</tr>
</tbody>
</table>
<p><strong>Variable type: numeric</strong></p>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 17%">
<col style="width: 12%">
<col style="width: 17%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 7%">
</colgroup>
<thead><tr class="header">
<th style="text-align: left;">skim_variable</th>
<th style="text-align: right;">n_missing</th>
<th style="text-align: right;">complete_rate</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">p0</th>
<th style="text-align: right;">p25</th>
<th style="text-align: right;">p50</th>
<th style="text-align: right;">p75</th>
<th style="text-align: right;">p100</th>
<th style="text-align: left;">hist</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">rate</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.18</td>
<td style="text-align: right;">0.09</td>
<td style="text-align: right;">0.05</td>
<td style="text-align: right;">0.10</td>
<td style="text-align: right;">0.2</td>
<td style="text-align: right;">0.25</td>
<td style="text-align: right;">0.30</td>
<td style="text-align: left;">▇▅▅▅▅</td>
</tr>
<tr class="even">
<td style="text-align: left;">expression</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">1</td>
<td style="text-align: right;">0.00</td>
<td style="text-align: right;">0.67</td>
<td style="text-align: right;">-6.50</td>
<td style="text-align: right;">-0.29</td>
<td style="text-align: right;">0.0</td>
<td style="text-align: right;">0.29</td>
<td style="text-align: right;">6.64</td>
<td style="text-align: left;">▁▁▇▁▁</td>
</tr>
</tbody>
</table>
</div>
</div>
</section></section><section id="accessing-variables-subsetting-data-frames" class="level2" data-number="2.4"><h2 data-number="2.4" class="anchored" data-anchor-id="accessing-variables-subsetting-data-frames">
<span class="header-section-number">2.4</span> Accessing variables &amp; subsetting data frames</h2>
<p>We can access individual variables within a data frame using the <code>$</code> operator, e.g., <code>mydataframe$specificVariable</code>. Let’s print out all the gene names in the data. Then let’s calculate the average expression across all conditions, all genes (using the built-in <code><a href="https://rdrr.io/r/base/mean.html">mean()</a></code> function).</p>
<div class="cell" data-hash="tibbles_cache/html/unnamed-chunk-4_77c940a967015d338a5d0740f64f5d05">
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># display all gene symbols</span></span>
<span><span class="va">ydat</span><span class="op">$</span><span class="va">symbol</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  [1] "SFB2"     NA         "QRI7"     "CFT2"     "SSO2"     "PSP2"    
  [7] "RIB2"     "VMA13"    "EDC3"     "VPS5"     NA         "AMN1"    
 [13] "SCW11"    "DSE2"     "COX15"    "SPE1"     "MTF1"     "KSS1"    
 [19] NA         NA         "YAP7"     NA         "YVC1"     "CDC40"   
 [25] NA         "RMD1"     "PCL6"     "AI4"      "GGC1"     "SUL1"    
 [31] "RAD57"    NA         "PER1"     "YHC3"     "SGE1"     "HNM1"    
 [37] "SWI1"     "NAM8"     NA         "BGL2"     "ACT1"     NA        
 [43] "SFL1"     "OYE3"     "MMP1"     "MHT1"     "SUL2"     "IPP1"    
 [49] "CWP1"     "SNF11"    "PEX25"    "ELO1"     NA         "CDC13"   
 [55] "FKH1"     "SWD1"     NA         "HOF1"     "HOC1"     "BNI5"    
 [61] "CSN12"    "PGS1"     "MLP2"     "HRP1"     NA         "SEC39"   
 [67] "ECM31"    NA         NA         "ADE4"     "ABC1"     "DLD2"    
 [73] "PHA2"     NA         "HAP3"     "MRPL23"   NA         NA        
 [79] "MRPL16"   NA         NA         NA         NA         "AI3"     
 [85] "COX1"     NA         "VAR1"     "COX3"     "COX2"     "AI5_BETA"
 [91] "AI2"      NA         NA         "GPI18"    "COS9"     NA        
 [97] NA         "PRP46"    "XDJ1"     "SLG1"     "MAM3"     "AEP1"    
[103] "UGO1"     NA         "RSC2"     "YAP1801"  "ZPR1"     "BCD1"    
[109] "UBP10"    "SLD3"     "RLF2"     "LRO1"     NA         "ITR2"    
[115] "ABP140"   "STT3"     "PTC2"     "STE20"    "HRD3"     "CWH43"   
[121] "ASK10"    "MPE1"     "SWC3"     "TSA1"     "ADE17"    "GFD2"    
[127] "PXR1"     NA         "BUD14"    "AUS1"     "NHX1"     "NTE1"    
[133] NA         "KIN3"     "BUD4"     "SLI15"    "PMT4"     "AVT5"    
[139] "CHS2"     "GPI13"    "KAP95"    "EFT2"     "EFT1"     "GAS1"    
[145] "CYK3"     "COQ2"     "PSD1"     NA         "PAC1"     "SUR7"    
[151] "RAX1"     "DFM1"     "RBD2"     NA         "YIP4"     "SRB2"    
[157] "HOL1"     "MEP3"     NA         "FEN2"     NA         "RFT1"    
[163] NA         "MCK1"     "GPI10"    "APT1"     NA         NA        
[169] "CPT1"     "ERV29"    "SFK1"     NA         "SEC20"    "TIR4"    
[175] NA         NA         "ARC35"    "SOL1"     "BIO2"     "ASC1"    
[181] "RBG1"     "PTC4"     NA         "OXA1"     "SIT4"     "PUB1"    
[187] "FPR4"     "FUN12"    "DPH2"     "DPS1"     "DLD1"     "ASN2"    
[193] "TRM9"     "DED81"    "SRM1"     "SAM50"    "POP2"     "FAA4"    
[199] NA         "CEM1"    
 [ reached getOption("max.print") -- omitted 198230 entries ]</code></pre>
</div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co">#mean expression</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">ydat</span><span class="op">$</span><span class="va">expression</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.00337</code></pre>
</div>
</div>
<p>Now that’s not too interesting. This is the average gene expression across all genes, across all conditions. The data is actually scaled/centered around zero:</p>
<div class="cell" data-hash="tibbles_cache/html/histogram_expression_values_740259c2289429ee2b8a652ebcba88db">
<div class="cell-output-display">
<p><img src="tibbles_files/figure-html/histogram_expression_values-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>We might be interested in the average expression of genes with a particular biological function, and how that changes over different growth rates restricted by particular nutrients. This is the kind of thing we’re going to do in the next section.</p>
<div class="callout callout-style-default callout-note no-icon callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon no-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise 1
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>What’s the standard deviation expression (hint: get help on the <code>sd</code> function with <code><a href="https://rdrr.io/r/stats/sd.html">?sd</a></code>).</li>
<li>What’s the range of rate represented in the data? (hint: <code><a href="https://rdrr.io/r/base/range.html">range()</a></code>).</li>
</ol>
</div>
</div>
</section><section id="bonus-preview-to-advanced-manipulation" class="level2" data-number="2.5"><h2 data-number="2.5" class="anchored" data-anchor-id="bonus-preview-to-advanced-manipulation">
<span class="header-section-number">2.5</span> BONUS: Preview to advanced manipulation</h2>
<p>What if we wanted show the mean expression, standard deviation, and correlation between growth rate and expression, separately for each limiting nutrient, separately for each gene, for all genes involved in the leucine biosynthesis pathway?</p>
<div class="cell" data-hash="tibbles_cache/html/unnamed-chunk-5_a21b9aa7ac253960553f3c7d30fffbba">
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">ydat</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">bp</span><span class="op">==</span><span class="st">"leucine biosynthesis"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">nutrient</span>, <span class="va">symbol</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarize</a></span><span class="op">(</span>mean<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">expression</span><span class="op">)</span>, sd<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/sd.html">sd</a></span><span class="op">(</span><span class="va">expression</span><span class="op">)</span>, r<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html">cor</a></span><span class="op">(</span><span class="va">rate</span>, <span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="tibbles_cache/html/unnamed-chunk-6_51d3341099ce3ad410e04e7891776a72">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead><tr class="header">
<th style="text-align: left;">nutrient</th>
<th style="text-align: left;">symbol</th>
<th style="text-align: right;">mean</th>
<th style="text-align: right;">sd</th>
<th style="text-align: right;">r</th>
</tr></thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Ammonia</td>
<td style="text-align: left;">LEU1</td>
<td style="text-align: right;">-0.82</td>
<td style="text-align: right;">0.39</td>
<td style="text-align: right;">0.66</td>
</tr>
<tr class="even">
<td style="text-align: left;">Ammonia</td>
<td style="text-align: left;">LEU2</td>
<td style="text-align: right;">-0.54</td>
<td style="text-align: right;">0.38</td>
<td style="text-align: right;">-0.19</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Ammonia</td>
<td style="text-align: left;">LEU4</td>
<td style="text-align: right;">-0.37</td>
<td style="text-align: right;">0.56</td>
<td style="text-align: right;">-0.67</td>
</tr>
<tr class="even">
<td style="text-align: left;">Ammonia</td>
<td style="text-align: left;">LEU9</td>
<td style="text-align: right;">-1.01</td>
<td style="text-align: right;">0.64</td>
<td style="text-align: right;">0.87</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Glucose</td>
<td style="text-align: left;">LEU1</td>
<td style="text-align: right;">-0.55</td>
<td style="text-align: right;">0.41</td>
<td style="text-align: right;">0.98</td>
</tr>
<tr class="even">
<td style="text-align: left;">Glucose</td>
<td style="text-align: left;">LEU2</td>
<td style="text-align: right;">-0.39</td>
<td style="text-align: right;">0.33</td>
<td style="text-align: right;">0.90</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Glucose</td>
<td style="text-align: left;">LEU4</td>
<td style="text-align: right;">1.09</td>
<td style="text-align: right;">1.01</td>
<td style="text-align: right;">-0.97</td>
</tr>
<tr class="even">
<td style="text-align: left;">Glucose</td>
<td style="text-align: left;">LEU9</td>
<td style="text-align: right;">-0.17</td>
<td style="text-align: right;">0.35</td>
<td style="text-align: right;">0.35</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Leucine</td>
<td style="text-align: left;">LEU1</td>
<td style="text-align: right;">2.70</td>
<td style="text-align: right;">1.08</td>
<td style="text-align: right;">-0.95</td>
</tr>
<tr class="even">
<td style="text-align: left;">Leucine</td>
<td style="text-align: left;">LEU2</td>
<td style="text-align: right;">0.28</td>
<td style="text-align: right;">1.16</td>
<td style="text-align: right;">-0.97</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Leucine</td>
<td style="text-align: left;">LEU4</td>
<td style="text-align: right;">0.80</td>
<td style="text-align: right;">1.06</td>
<td style="text-align: right;">-0.97</td>
</tr>
<tr class="even">
<td style="text-align: left;">Leucine</td>
<td style="text-align: left;">LEU9</td>
<td style="text-align: right;">0.39</td>
<td style="text-align: right;">0.18</td>
<td style="text-align: right;">-0.77</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Phosphate</td>
<td style="text-align: left;">LEU1</td>
<td style="text-align: right;">-0.43</td>
<td style="text-align: right;">0.27</td>
<td style="text-align: right;">0.95</td>
</tr>
<tr class="even">
<td style="text-align: left;">Phosphate</td>
<td style="text-align: left;">LEU2</td>
<td style="text-align: right;">-0.26</td>
<td style="text-align: right;">0.19</td>
<td style="text-align: right;">0.70</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Phosphate</td>
<td style="text-align: left;">LEU4</td>
<td style="text-align: right;">-0.99</td>
<td style="text-align: right;">0.11</td>
<td style="text-align: right;">0.24</td>
</tr>
<tr class="even">
<td style="text-align: left;">Phosphate</td>
<td style="text-align: left;">LEU9</td>
<td style="text-align: right;">-1.12</td>
<td style="text-align: right;">0.53</td>
<td style="text-align: right;">0.90</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Sulfate</td>
<td style="text-align: left;">LEU1</td>
<td style="text-align: right;">-1.17</td>
<td style="text-align: right;">0.34</td>
<td style="text-align: right;">0.98</td>
</tr>
<tr class="even">
<td style="text-align: left;">Sulfate</td>
<td style="text-align: left;">LEU2</td>
<td style="text-align: right;">-0.96</td>
<td style="text-align: right;">0.30</td>
<td style="text-align: right;">0.57</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Sulfate</td>
<td style="text-align: left;">LEU4</td>
<td style="text-align: right;">-0.24</td>
<td style="text-align: right;">0.43</td>
<td style="text-align: right;">-0.60</td>
</tr>
<tr class="even">
<td style="text-align: left;">Sulfate</td>
<td style="text-align: left;">LEU9</td>
<td style="text-align: right;">-1.24</td>
<td style="text-align: right;">0.55</td>
<td style="text-align: right;">0.99</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Uracil</td>
<td style="text-align: left;">LEU1</td>
<td style="text-align: right;">-0.74</td>
<td style="text-align: right;">0.73</td>
<td style="text-align: right;">0.89</td>
</tr>
<tr class="even">
<td style="text-align: left;">Uracil</td>
<td style="text-align: left;">LEU2</td>
<td style="text-align: right;">0.18</td>
<td style="text-align: right;">0.13</td>
<td style="text-align: right;">-0.07</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Uracil</td>
<td style="text-align: left;">LEU4</td>
<td style="text-align: right;">-0.65</td>
<td style="text-align: right;">0.44</td>
<td style="text-align: right;">0.77</td>
</tr>
<tr class="even">
<td style="text-align: left;">Uracil</td>
<td style="text-align: left;">LEU9</td>
<td style="text-align: right;">-1.02</td>
<td style="text-align: right;">0.91</td>
<td style="text-align: right;">0.94</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Neat eh? We’ll learn how to do that in the advanced manipulation with dplyr chapter.</p>
<!-- Neat eh? Let's learn how to do that in the [advanced manipulation with dplyr chapter](r-dplyr-yeast.html). -->


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><nav class="page-navigation"><div class="nav-page nav-page-previous">
      <a href="./basics.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Basics</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./dplyr.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation with dplyr</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">© 2019 Stephen D. Turner, Ph.D.</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
<li class="nav-item compact">
    <a class="nav-link" href="https://x.com/strnr">
      <i class="bi bi-twitter" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/stephenturner">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/turnersd/">
      <i class="bi bi-linkedin" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://stephenturner.us/">
      <i class="bi bi-house-fill" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://blog.stephenturner.us/">
      <i class="bi bi-rss" role="img">
</i> 
    </a>
  </li>  
</ul>
</div>
  </div>
</footer>


</body></html>