<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Biological Data Science with R - 4&nbsp; Tidy Data and Advanced Data Manipulation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./ggplot2.html" rel="next">
<link href="./dplyr.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./basics.html">Core lessons</a></li><li class="breadcrumb-item"><a href="./tidyr.html"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Tidy Data and Advanced Data Manipulation</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Biological Data Science with R</a> 
        <div class="sidebar-tools-main">
    <div class="dropdown">
      <a href="" title="Download" id="quarto-navigation-tool-dropdown-0" class="quarto-navigation-tool dropdown-toggle px-1" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download"><i class="bi bi-download"></i></a>
      <ul class="dropdown-menu" aria-labelledby="quarto-navigation-tool-dropdown-0">
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Biological-Data-Science-with-R.pdf">
              <i class="bi bi-bi-file-pdf pe-1"></i>
            Download PDF
            </a>
          </li>
          <li>
            <a class="dropdown-item sidebar-tools-main-item" href="./Biological-Data-Science-with-R.epub">
              <i class="bi bi-bi-journal pe-1"></i>
            Download ePub
            </a>
          </li>
      </ul>
    </div>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Core lessons</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Basics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tibbles.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Tibbles</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./dplyr.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation with dplyr</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tidyr.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Tidy Data and Advanced Data Manipulation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ggplot2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Visualization with ggplot2</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./tidyeda.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Refresher: Tidy Exploratory Data Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./rmarkdown.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Reproducible Reporting with RMarkdown</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Electives</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./stats.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Essential Statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./survival.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Survival Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./predmodeling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Predictive Modeling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./forecasting.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Probabilistic Forecasting</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./textmining.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Text Mining</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ggtree.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Phylogenetic Trees</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./rnaseq.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">RNA-seq</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summary.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Summary</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./setup.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">A</span>&nbsp; <span class="chapter-title">Setup</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">B</span>&nbsp; <span class="chapter-title">Additional Resources</span></span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#tidy-data" id="toc-tidy-data" class="nav-link active" data-scroll-target="#tidy-data"><span class="header-section-number">4.1</span> Tidy data</a></li>
  <li><a href="#the-tidyr-package" id="toc-the-tidyr-package" class="nav-link" data-scroll-target="#the-tidyr-package"><span class="header-section-number">4.2</span> The <strong>tidyr</strong> package</a>
  <ul class="collapse">
  <li><a href="#gather" id="toc-gather" class="nav-link" data-scroll-target="#gather"><span class="header-section-number">4.2.1</span> <code>gather()</code></a></li>
  <li><a href="#separate" id="toc-separate" class="nav-link" data-scroll-target="#separate"><span class="header-section-number">4.2.2</span> <code>separate()</code></a></li>
  <li><a href="#it-all-together" id="toc-it-all-together" class="nav-link" data-scroll-target="#it-all-together"><span class="header-section-number">4.2.3</span> <code>|&gt;</code> it all together</a></li>
  </ul></li>
  <li><a href="#tidy-the-yeast-data" id="toc-tidy-the-yeast-data" class="nav-link" data-scroll-target="#tidy-the-yeast-data"><span class="header-section-number">4.3</span> Tidy the yeast data</a>
  <ul class="collapse">
  <li><a href="#separate-the-name" id="toc-separate-the-name" class="nav-link" data-scroll-target="#separate-the-name"><span class="header-section-number">4.3.1</span> <code>separate()</code> the <code>NAME</code></a></li>
  <li><a href="#gather-the-data" id="toc-gather-the-data" class="nav-link" data-scroll-target="#gather-the-data"><span class="header-section-number">4.3.2</span> <code>gather()</code> the data</a></li>
  <li><a href="#inner_join-to-go" id="toc-inner_join-to-go" class="nav-link" data-scroll-target="#inner_join-to-go"><span class="header-section-number">4.3.3</span> <code>inner_join()</code> to GO</a></li>
  <li><a href="#finishing-touches" id="toc-finishing-touches" class="nav-link" data-scroll-target="#finishing-touches"><span class="header-section-number">4.3.4</span> Finishing touches</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="sec-tidyr" class="quarto-section-identifier"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Tidy Data and Advanced Data Manipulation</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><strong>Recommended reading prior to class</strong>: Sections 1-3 of <a href="http://www.jstatsoft.org/article/view/v059i10/v59i10.pdf">Wickham, H. “Tidy Data.” <em>Journal of Statistical Software</em> 59:10 (2014)</a>.</p>
<p>Data needed:</p>
<ul>
<li>Heart rate data: <a href="data/heartrate2dose.csv">heartrate2dose.csv</a></li>
<li><em>Tidy</em> yeast data: <a href="data/brauer2007_tidy.csv">brauer2007_tidy.csv</a></li>
<li><em>Original</em> (untidy) yeast data: <a href="data/brauer2007_messy.csv">brauer2007_messy.csv</a></li>
<li>Yeast systematic names to GO terms: <a href="data/brauer2007_sysname2go.csv">brauer2007_sysname2go.csv</a></li>
</ul>
<section id="tidy-data" class="level2" data-number="4.1">
<h2 data-number="4.1" class="anchored" data-anchor-id="tidy-data"><span class="header-section-number">4.1</span> Tidy data</h2>
<p>So far we’ve dealt exclusively with tidy data – data that’s easy to work with, manipulate, and visualize. That’s because our dataset has two key properties:</p>
<ol type="1">
<li>Each <em>column</em> is a <em>variable</em>.</li>
<li>Each <em>row</em> is an <em>observation</em>.</li>
</ol>
<p>You can read a lot more about tidy data <a href="http://www.jstatsoft.org/v59/i10/paper">in this paper</a>. Let’s load some untidy data and see if we can see the difference. This is some made-up data for five different patients (Jon, Ann, Bill, Kate, and Joe) given three different drugs (A, B, and C), at two doses (10 and 20), and measuring their heart rate. Download the <a href="data/heartrate2dose.csv">heartrate2dose.csv</a> file. Load <strong>readr</strong> and <strong>dplyr</strong>, and import and display the data.</p>
<div class="cell" data-hash="tidyr_cache/html/untidy_1d5837ff1933698b7d7d918dd4b7474e">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>hr <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/heartrate2dose.csv"</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>hr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 7
  name   a_10  a_20  b_10  b_20  c_10  c_20
  &lt;chr&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1 jon      60    55    65    60    70    70
2 ann      65    60    70    65    75    75
3 bill     70    65    75    70    80    80
4 kate     75    70    80    75    85    85
5 joe      80    75    85    80    90    90</code></pre>
</div>
</div>
<p>Notice how with the yeast data each variable (symbol, nutrient, rate, expression, etc.) were each in their own column. In this heart rate data, we have four variables: name, drug, dose, and heart rate. <em>Name</em> is in a column, but <em>drug</em> is in the header row. Furthermore the drug and <em>dose</em> are tied together in the same column, and the <em>heart rate</em> is scattered around the entire table. If we wanted to do things like <code>filter</code> the dataset where <code>drug=="a"</code> or <code>dose==20</code> or <code>heartrate&gt;=80</code> we couldn’t do it because these variables aren’t in columns.</p>
</section>
<section id="the-tidyr-package" class="level2" data-number="4.2">
<h2 data-number="4.2" class="anchored" data-anchor-id="the-tidyr-package"><span class="header-section-number">4.2</span> The <strong>tidyr</strong> package</h2>
<p>The <strong>tidyr</strong> package helps with this. There are several functions in the tidyr package but the ones we’re going to use are <code>separate()</code> and <code>gather()</code>. The <code>gather()</code> function takes multiple columns, and gathers them into key-value pairs: it makes “wide” data longer. The <code>separate()</code> function separates one column into multiple columns. So, what we need to do is <em>gather</em> all the drug/dose data into a column with their corresponding heart rate, and then <em>separate</em> that column into two separate columns for the drug and dose.</p>
<p>Before we get started, load the <strong>tidyr</strong> package, and look at the help pages for <code>?gather</code> and <code>?separate</code>. Notice how each of these functions takes a data frame as input and returns a data frame as output. Thus, we can pipe from one function to the next.</p>
<div class="cell" data-hash="tidyr_cache/html/tidydata_7b4771cd98caa45a535bb5c80e771f87">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="gather" class="level3" data-number="4.2.1">
<h3 data-number="4.2.1" class="anchored" data-anchor-id="gather"><span class="header-section-number">4.2.1</span> <code>gather()</code></h3>
<p>The help for <code>?gather</code> tells us that we first pass in a data frame (or omit the first argument, and pipe in the data with <code>|&gt;</code>). The next two arguments are the names of the key and value columns to create, and all the relevant arguments that come after that are the columns we want to <em>gather</em> together. Here’s one way to do it.</p>
<div class="cell" data-hash="tidyr_cache/html/gather1_ee8b59ee563f94442d2300455db55f22">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>hr <span class="sc">|&gt;</span> <span class="fu">gather</span>(<span class="at">key=</span>drugdose, <span class="at">value=</span>hr, a_10, a_20, b_10, b_20, c_10, c_20)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 3
   name  drugdose    hr
   &lt;chr&gt; &lt;chr&gt;    &lt;dbl&gt;
 1 jon   a_10        60
 2 ann   a_10        65
 3 bill  a_10        70
 4 kate  a_10        75
 5 joe   a_10        80
 6 jon   a_20        55
 7 ann   a_20        60
 8 bill  a_20        65
 9 kate  a_20        70
10 joe   a_20        75
# ℹ 20 more rows</code></pre>
</div>
</div>
<p>But that gets cumbersome to type all those names. What if we had 100 drugs and 3 doses of each? There are two other ways of specifying which columns to gather. The help for <code>?gather</code> tells you how to do this:</p>
<blockquote class="blockquote">
<p><code>...</code> Specification of columns to gather. Use bare variable names. Select all variables between x and z with x:z, exclude y with -y. For more options, see the <code>select</code> documentation.</p>
</blockquote>
<p>So, we could accomplish the same thing by doing this:</p>
<div class="cell" data-hash="tidyr_cache/html/gather2_9165b675cdec0ab5694c4d570cb8b184">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>hr <span class="sc">|&gt;</span> <span class="fu">gather</span>(<span class="at">key=</span>drugdose, <span class="at">value=</span>hr, a_10<span class="sc">:</span>c_20)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 3
   name  drugdose    hr
   &lt;chr&gt; &lt;chr&gt;    &lt;dbl&gt;
 1 jon   a_10        60
 2 ann   a_10        65
 3 bill  a_10        70
 4 kate  a_10        75
 5 joe   a_10        80
 6 jon   a_20        55
 7 ann   a_20        60
 8 bill  a_20        65
 9 kate  a_20        70
10 joe   a_20        75
# ℹ 20 more rows</code></pre>
</div>
</div>
<p>But what if we didn’t know the drug names or doses, but we <em>did</em> know that the only other column in there that we <em>don’t</em> want to gather is <code>name</code>?</p>
<div class="cell" data-hash="tidyr_cache/html/gather3_ed8a660184bbeda69dd1e348e0aa7a0a">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>hr <span class="sc">|&gt;</span> <span class="fu">gather</span>(<span class="at">key=</span>drugdose, <span class="at">value=</span>hr, <span class="sc">-</span>name)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 3
   name  drugdose    hr
   &lt;chr&gt; &lt;chr&gt;    &lt;dbl&gt;
 1 jon   a_10        60
 2 ann   a_10        65
 3 bill  a_10        70
 4 kate  a_10        75
 5 joe   a_10        80
 6 jon   a_20        55
 7 ann   a_20        60
 8 bill  a_20        65
 9 kate  a_20        70
10 joe   a_20        75
# ℹ 20 more rows</code></pre>
</div>
</div>
</section>
<section id="separate" class="level3" data-number="4.2.2">
<h3 data-number="4.2.2" class="anchored" data-anchor-id="separate"><span class="header-section-number">4.2.2</span> <code>separate()</code></h3>
<p>Finally, look at the help for <code>?separate</code>. We can pipe in data and omit the first argument. The second argument is the column to separate; the <code>into</code> argument is a <em>character vector</em> of the new column names, and the <code>sep</code> argument is a character used to separate columns, or a number indicating the position to split at.</p>
<blockquote class="blockquote">
<p><strong><em>Side note, and 60-second lesson on vectors</em></strong>: We can create arbitrary-length <em>vectors</em>, which are simply variables that contain an arbitrary number of values. To create a numeric vector, try this: <code>c(5, 42, 22908)</code>. That creates a three element vector. Try <code>c("cat", "dog")</code>.</p>
</blockquote>
<div class="cell" data-hash="tidyr_cache/html/gathersep_5b9686ae07c14a31f0dbd5c878309950">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>hr <span class="sc">|&gt;</span> </span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gather</span>(<span class="at">key=</span>drugdose, <span class="at">value=</span>hr, <span class="sc">-</span>name) <span class="sc">|&gt;</span> </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(drugdose, <span class="at">into=</span><span class="fu">c</span>(<span class="st">"drug"</span>, <span class="st">"dose"</span>), <span class="at">sep=</span><span class="st">"_"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 30 × 4
   name  drug  dose     hr
   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;
 1 jon   a     10       60
 2 ann   a     10       65
 3 bill  a     10       70
 4 kate  a     10       75
 5 joe   a     10       80
 6 jon   a     20       55
 7 ann   a     20       60
 8 bill  a     20       65
 9 kate  a     20       70
10 joe   a     20       75
# ℹ 20 more rows</code></pre>
</div>
</div>
</section>
<section id="it-all-together" class="level3" data-number="4.2.3">
<h3 data-number="4.2.3" class="anchored" data-anchor-id="it-all-together"><span class="header-section-number">4.2.3</span> <code>|&gt;</code> it all together</h3>
<p>Let’s put it all together with <code>gather |&gt; separate |&gt; filter |&gt; group_by |&gt; summarize</code>.</p>
<p>If we create a new data frame that’s a tidy version of hr, we can do those kinds of manipulations we talked about before:</p>
<div class="cell" data-hash="tidyr_cache/html/hrtidy_3effe555bf390f2bec7f6c81972858da">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a new data.frame</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>hrtidy <span class="ot">&lt;-</span> hr <span class="sc">|&gt;</span> </span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gather</span>(<span class="at">key=</span>drugdose, <span class="at">value=</span>hr, <span class="sc">-</span>name) <span class="sc">|&gt;</span> </span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(drugdose, <span class="at">into=</span><span class="fu">c</span>(<span class="st">"drug"</span>, <span class="st">"dose"</span>), <span class="at">sep=</span><span class="st">"_"</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Optionally, view it</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a><span class="co"># View(hrtidy)</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="co"># filter</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>hrtidy <span class="sc">|&gt;</span> <span class="fu">filter</span>(drug<span class="sc">==</span><span class="st">"a"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 4
   name  drug  dose     hr
   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;
 1 jon   a     10       60
 2 ann   a     10       65
 3 bill  a     10       70
 4 kate  a     10       75
 5 joe   a     10       80
 6 jon   a     20       55
 7 ann   a     20       60
 8 bill  a     20       65
 9 kate  a     20       70
10 joe   a     20       75</code></pre>
</div>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>hrtidy <span class="sc">|&gt;</span> <span class="fu">filter</span>(dose<span class="sc">==</span><span class="dv">20</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 15 × 4
   name  drug  dose     hr
   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;
 1 jon   a     20       55
 2 ann   a     20       60
 3 bill  a     20       65
 4 kate  a     20       70
 5 joe   a     20       75
 6 jon   b     20       60
 7 ann   b     20       65
 8 bill  b     20       70
 9 kate  b     20       75
10 joe   b     20       80
11 jon   c     20       70
12 ann   c     20       75
13 bill  c     20       80
14 kate  c     20       85
15 joe   c     20       90</code></pre>
</div>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>hrtidy <span class="sc">|&gt;</span> <span class="fu">filter</span>(hr<span class="sc">&gt;=</span><span class="dv">80</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 4
   name  drug  dose     hr
   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;dbl&gt;
 1 joe   a     10       80
 2 kate  b     10       80
 3 joe   b     10       85
 4 joe   b     20       80
 5 bill  c     10       80
 6 kate  c     10       85
 7 joe   c     10       90
 8 bill  c     20       80
 9 kate  c     20       85
10 joe   c     20       90</code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># analyze</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>hrtidy <span class="sc">|&gt;</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(name<span class="sc">!=</span><span class="st">"joe"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(drug, dose) <span class="sc">|&gt;</span></span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">meanhr=</span><span class="fu">mean</span>(hr))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
# Groups:   drug [3]
  drug  dose  meanhr
  &lt;chr&gt; &lt;chr&gt;  &lt;dbl&gt;
1 a     10      67.5
2 a     20      62.5
3 b     10      72.5
4 b     20      67.5
5 c     10      77.5
6 c     20      77.5</code></pre>
</div>
</div>
</section>
</section>
<section id="tidy-the-yeast-data" class="level2" data-number="4.3">
<h2 data-number="4.3" class="anchored" data-anchor-id="tidy-the-yeast-data"><span class="header-section-number">4.3</span> Tidy the yeast data</h2>
<p>Now, let’s take a look at the yeast data again. The data we’ve been working with up to this point was already cleaned up to a good degree. All of our variables (symbol, nutrient, rate, expression, GO terms, etc.) were each in their own column. Make sure you have the necessary libraries loaded, and read in the tidy data once more into an object called <code>ydat</code>.</p>
<div class="cell" data-hash="tidyr_cache/html/unnamed-chunk-2_b2855e5abb86e11e2d1a010ca245a904">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(readr)</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Import data</span></span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>ydat <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/brauer2007_tidy.csv"</span>)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Optionally, View</span></span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a><span class="co"># View(ydat)</span></span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Or just display to the screen</span></span>
<span id="cb20-13"><a href="#cb20-13" aria-hidden="true" tabindex="-1"></a>ydat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 198,430 × 7
   symbol systematic_name nutrient  rate expression bp                     mf   
   &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;    &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;                  &lt;chr&gt;
 1 SFB2   YNL049C         Glucose   0.05      -0.24 ER to Golgi transport  mole…
 2 &lt;NA&gt;   YNL095C         Glucose   0.05       0.28 biological process un… mole…
 3 QRI7   YDL104C         Glucose   0.05      -0.02 proteolysis and pepti… meta…
 4 CFT2   YLR115W         Glucose   0.05      -0.33 mRNA polyadenylylatio… RNA …
 5 SSO2   YMR183C         Glucose   0.05       0.05 vesicle fusion*        t-SN…
 6 PSP2   YML017W         Glucose   0.05      -0.69 biological process un… mole…
 7 RIB2   YOL066C         Glucose   0.05      -0.55 riboflavin biosynthes… pseu…
 8 VMA13  YPR036W         Glucose   0.05      -0.75 vacuolar acidification hydr…
 9 EDC3   YEL015W         Glucose   0.05      -0.24 deadenylylation-indep… mole…
10 VPS5   YOR069W         Glucose   0.05      -0.16 protein retention in … prot…
# ℹ 198,420 more rows</code></pre>
</div>
</div>
<p>But let’s take a look to see what this data originally looked like.</p>
<div class="cell" data-hash="tidyr_cache/html/unnamed-chunk-3_eb8b01d2fe01e4c5b55e4bea222a4135">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>yorig <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/brauer2007_messy.csv"</span>)</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># View(yorig)</span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>yorig</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5,536 × 40
   GID       YORF  NAME  GWEIGHT G0.05  G0.1 G0.15  G0.2 G0.25  G0.3 N0.05  N0.1
   &lt;chr&gt;     &lt;chr&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 GENE1331X A_06… SFB2…       1 -0.24 -0.13 -0.21 -0.15 -0.05 -0.05  0.2   0.24
 2 GENE4924X A_06… NA::…       1  0.28  0.13 -0.4  -0.48 -0.11  0.17  0.31  0   
 3 GENE4690X A_06… QRI7…       1 -0.02 -0.27 -0.27 -0.02  0.24  0.25  0.23  0.06
 4 GENE1177X A_06… CFT2…       1 -0.33 -0.41 -0.24 -0.03 -0.03  0     0.2  -0.25
 5 GENE511X  A_06… SSO2…       1  0.05  0.02  0.4   0.34 -0.13 -0.14 -0.35 -0.09
 6 GENE2133X A_06… PSP2…       1 -0.69 -0.03  0.23  0.2   0    -0.27  0.17 -0.4 
 7 GENE1002X A_06… RIB2…       1 -0.55 -0.3  -0.12 -0.03 -0.16 -0.11  0.04  0   
 8 GENE5478X A_06… VMA1…       1 -0.75 -0.12 -0.07  0.02 -0.32 -0.41  0.11 -0.16
 9 GENE2065X A_06… EDC3…       1 -0.24 -0.22  0.14  0.06  0    -0.13  0.3   0.07
10 GENE2440X A_06… VPS5…       1 -0.16 -0.38  0.05  0.14 -0.04 -0.01  0.39  0.2 
# ℹ 5,526 more rows
# ℹ 28 more variables: N0.15 &lt;dbl&gt;, N0.2 &lt;dbl&gt;, N0.25 &lt;dbl&gt;, N0.3 &lt;dbl&gt;,
#   P0.05 &lt;dbl&gt;, P0.1 &lt;dbl&gt;, P0.15 &lt;dbl&gt;, P0.2 &lt;dbl&gt;, P0.25 &lt;dbl&gt;, P0.3 &lt;dbl&gt;,
#   S0.05 &lt;dbl&gt;, S0.1 &lt;dbl&gt;, S0.15 &lt;dbl&gt;, S0.2 &lt;dbl&gt;, S0.25 &lt;dbl&gt;, S0.3 &lt;dbl&gt;,
#   L0.05 &lt;dbl&gt;, L0.1 &lt;dbl&gt;, L0.15 &lt;dbl&gt;, L0.2 &lt;dbl&gt;, L0.25 &lt;dbl&gt;, L0.3 &lt;dbl&gt;,
#   U0.05 &lt;dbl&gt;, U0.1 &lt;dbl&gt;, U0.15 &lt;dbl&gt;, U0.2 &lt;dbl&gt;, U0.25 &lt;dbl&gt;, U0.3 &lt;dbl&gt;</code></pre>
</div>
</div>
<p>There are several issues here.</p>
<ol type="1">
<li><strong>Multiple variables are stored in one column.</strong> The <code>NAME</code> column contains lots of information, split up by <code>::</code>’s.</li>
<li><strong>Nutrient and rate variables are stuck in column headers.</strong> That is, the column names contain the values of two variables: nutrient (G, N, P, S, L, U) and growth rate (0.05-0.3). Remember, with tidy data, <strong>each <em>column</em> is a <em>variable</em> and each <em>row</em> is an <em>observation</em>.</strong> Here, we have not one observation per row, but 36 (6 nutrients <span class="math inline">\(\times\)</span> 6 rates)! There’s no way we could filter this data by a certain nutrient, or try to calculate statistics between rate and expression.</li>
<li><strong>Expression values are scattered throughout the table.</strong> Related to the problem above, and just like our heart rate example, <code>expression</code> isn’t a single-column variable as in the cleaned tidy data, but it’s scattered around these 36 columns.</li>
<li><strong>Other important information is in a separate table.</strong> We’re missing all the gene ontology information we had in the tidy data (no information about biological process (<code>bp</code>) or molecular function (<code>mf</code>)).</li>
</ol>
<p>Let’s tackle these issues one at a time, all on a <code>|&gt;</code> pipeline.</p>
<section id="separate-the-name" class="level3" data-number="4.3.1">
<h3 data-number="4.3.1" class="anchored" data-anchor-id="separate-the-name"><span class="header-section-number">4.3.1</span> <code>separate()</code> the <code>NAME</code></h3>
<p>Let’s <code>separate()</code> the <code>NAME</code> column <code>into</code> multiple different variables. The first row looks like this:</p>
<blockquote class="blockquote">
<p><code>SFB2::YNL049C::1082129</code></p>
</blockquote>
<p>That is, it looks like we’ve got the gene symbol, the systematic name, and some other number (that isn’t discussed in the paper). Let’s <code>separate()</code>!</p>
<div class="cell" data-hash="tidyr_cache/html/unnamed-chunk-4_c76febefad296a359c3d10ee47e9b85b">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>yorig <span class="sc">|&gt;</span> </span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(NAME, <span class="at">into=</span><span class="fu">c</span>(<span class="st">"symbol"</span>, <span class="st">"systematic_name"</span>, <span class="st">"somenumber"</span>), <span class="at">sep=</span><span class="st">"::"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5,536 × 42
   GID   YORF  symbol systematic_name somenumber GWEIGHT G0.05  G0.1 G0.15  G0.2
   &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;        &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 GENE… A_06… SFB2   YNL049C         1082129          1 -0.24 -0.13 -0.21 -0.15
 2 GENE… A_06… NA     YNL095C         1086222          1  0.28  0.13 -0.4  -0.48
 3 GENE… A_06… QRI7   YDL104C         1085955          1 -0.02 -0.27 -0.27 -0.02
 4 GENE… A_06… CFT2   YLR115W         1081958          1 -0.33 -0.41 -0.24 -0.03
 5 GENE… A_06… SSO2   YMR183C         1081214          1  0.05  0.02  0.4   0.34
 6 GENE… A_06… PSP2   YML017W         1083036          1 -0.69 -0.03  0.23  0.2 
 7 GENE… A_06… RIB2   YOL066C         1081766          1 -0.55 -0.3  -0.12 -0.03
 8 GENE… A_06… VMA13  YPR036W         1086860          1 -0.75 -0.12 -0.07  0.02
 9 GENE… A_06… EDC3   YEL015W         1082963          1 -0.24 -0.22  0.14  0.06
10 GENE… A_06… VPS5   YOR069W         1083389          1 -0.16 -0.38  0.05  0.14
# ℹ 5,526 more rows
# ℹ 32 more variables: G0.25 &lt;dbl&gt;, G0.3 &lt;dbl&gt;, N0.05 &lt;dbl&gt;, N0.1 &lt;dbl&gt;,
#   N0.15 &lt;dbl&gt;, N0.2 &lt;dbl&gt;, N0.25 &lt;dbl&gt;, N0.3 &lt;dbl&gt;, P0.05 &lt;dbl&gt;, P0.1 &lt;dbl&gt;,
#   P0.15 &lt;dbl&gt;, P0.2 &lt;dbl&gt;, P0.25 &lt;dbl&gt;, P0.3 &lt;dbl&gt;, S0.05 &lt;dbl&gt;, S0.1 &lt;dbl&gt;,
#   S0.15 &lt;dbl&gt;, S0.2 &lt;dbl&gt;, S0.25 &lt;dbl&gt;, S0.3 &lt;dbl&gt;, L0.05 &lt;dbl&gt;, L0.1 &lt;dbl&gt;,
#   L0.15 &lt;dbl&gt;, L0.2 &lt;dbl&gt;, L0.25 &lt;dbl&gt;, L0.3 &lt;dbl&gt;, U0.05 &lt;dbl&gt;, U0.1 &lt;dbl&gt;,
#   U0.15 &lt;dbl&gt;, U0.2 &lt;dbl&gt;, U0.25 &lt;dbl&gt;, U0.3 &lt;dbl&gt;</code></pre>
</div>
</div>
<p>Now, let’s <code>select()</code> out the stuff we don’t want.</p>
<div class="cell" data-hash="tidyr_cache/html/unnamed-chunk-5_e9f94c722a6e0c5a0e1a13dbd902df4f">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>yorig <span class="sc">|&gt;</span> </span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(NAME, <span class="at">into=</span><span class="fu">c</span>(<span class="st">"symbol"</span>, <span class="st">"systematic_name"</span>, <span class="st">"somenumber"</span>), <span class="at">sep=</span><span class="st">"::"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>GID, <span class="sc">-</span>YORF, <span class="sc">-</span>somenumber, <span class="sc">-</span>GWEIGHT)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5,536 × 38
   symbol systematic_name G0.05  G0.1 G0.15  G0.2 G0.25  G0.3 N0.05  N0.1 N0.15
   &lt;chr&gt;  &lt;chr&gt;           &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 SFB2   YNL049C         -0.24 -0.13 -0.21 -0.15 -0.05 -0.05  0.2   0.24 -0.2 
 2 NA     YNL095C          0.28  0.13 -0.4  -0.48 -0.11  0.17  0.31  0    -0.63
 3 QRI7   YDL104C         -0.02 -0.27 -0.27 -0.02  0.24  0.25  0.23  0.06 -0.66
 4 CFT2   YLR115W         -0.33 -0.41 -0.24 -0.03 -0.03  0     0.2  -0.25 -0.49
 5 SSO2   YMR183C          0.05  0.02  0.4   0.34 -0.13 -0.14 -0.35 -0.09 -0.08
 6 PSP2   YML017W         -0.69 -0.03  0.23  0.2   0    -0.27  0.17 -0.4  -0.54
 7 RIB2   YOL066C         -0.55 -0.3  -0.12 -0.03 -0.16 -0.11  0.04  0    -0.63
 8 VMA13  YPR036W         -0.75 -0.12 -0.07  0.02 -0.32 -0.41  0.11 -0.16 -0.26
 9 EDC3   YEL015W         -0.24 -0.22  0.14  0.06  0    -0.13  0.3   0.07 -0.3 
10 VPS5   YOR069W         -0.16 -0.38  0.05  0.14 -0.04 -0.01  0.39  0.2   0.27
# ℹ 5,526 more rows
# ℹ 27 more variables: N0.2 &lt;dbl&gt;, N0.25 &lt;dbl&gt;, N0.3 &lt;dbl&gt;, P0.05 &lt;dbl&gt;,
#   P0.1 &lt;dbl&gt;, P0.15 &lt;dbl&gt;, P0.2 &lt;dbl&gt;, P0.25 &lt;dbl&gt;, P0.3 &lt;dbl&gt;, S0.05 &lt;dbl&gt;,
#   S0.1 &lt;dbl&gt;, S0.15 &lt;dbl&gt;, S0.2 &lt;dbl&gt;, S0.25 &lt;dbl&gt;, S0.3 &lt;dbl&gt;, L0.05 &lt;dbl&gt;,
#   L0.1 &lt;dbl&gt;, L0.15 &lt;dbl&gt;, L0.2 &lt;dbl&gt;, L0.25 &lt;dbl&gt;, L0.3 &lt;dbl&gt;, U0.05 &lt;dbl&gt;,
#   U0.1 &lt;dbl&gt;, U0.15 &lt;dbl&gt;, U0.2 &lt;dbl&gt;, U0.25 &lt;dbl&gt;, U0.3 &lt;dbl&gt;</code></pre>
</div>
</div>
</section>
<section id="gather-the-data" class="level3" data-number="4.3.2">
<h3 data-number="4.3.2" class="anchored" data-anchor-id="gather-the-data"><span class="header-section-number">4.3.2</span> <code>gather()</code> the data</h3>
<p>Let’s gather the data from wide to long format so we get nutrient/rate (key) and expression (value) in their own columns.</p>
<div class="cell" data-hash="tidyr_cache/html/unnamed-chunk-6_5995d5a05a2c936b5176405b043e1f02">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>yorig <span class="sc">|&gt;</span> </span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(NAME, <span class="at">into=</span><span class="fu">c</span>(<span class="st">"symbol"</span>, <span class="st">"systematic_name"</span>, <span class="st">"somenumber"</span>), <span class="at">sep=</span><span class="st">"::"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>GID, <span class="sc">-</span>YORF, <span class="sc">-</span>somenumber, <span class="sc">-</span>GWEIGHT) <span class="sc">|&gt;</span> </span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gather</span>(<span class="at">key=</span>nutrientrate, <span class="at">value=</span>expression, G0<span class="fl">.05</span><span class="sc">:</span>U0<span class="fl">.3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 199,296 × 4
   symbol systematic_name nutrientrate expression
   &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;             &lt;dbl&gt;
 1 SFB2   YNL049C         G0.05             -0.24
 2 NA     YNL095C         G0.05              0.28
 3 QRI7   YDL104C         G0.05             -0.02
 4 CFT2   YLR115W         G0.05             -0.33
 5 SSO2   YMR183C         G0.05              0.05
 6 PSP2   YML017W         G0.05             -0.69
 7 RIB2   YOL066C         G0.05             -0.55
 8 VMA13  YPR036W         G0.05             -0.75
 9 EDC3   YEL015W         G0.05             -0.24
10 VPS5   YOR069W         G0.05             -0.16
# ℹ 199,286 more rows</code></pre>
</div>
</div>
<p>And while we’re at it, let’s <code>separate()</code> that newly created key column. Take a look at the help for <code>?separate</code> again. The <code>sep</code> argument could be a delimiter or a number position to split at. Let’s split after the first character. While we’re at it, let’s hold onto this intermediate data frame before we add gene ontology information. Call it <code>ynogo</code>.</p>
<div class="cell" data-hash="tidyr_cache/html/unnamed-chunk-7_26f9e37c91e9f12c34e75a1edb9ede61">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>ynogo <span class="ot">&lt;-</span> yorig <span class="sc">|&gt;</span> </span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(NAME, <span class="at">into=</span><span class="fu">c</span>(<span class="st">"symbol"</span>, <span class="st">"systematic_name"</span>, <span class="st">"somenumber"</span>), <span class="at">sep=</span><span class="st">"::"</span>) <span class="sc">|&gt;</span> </span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>GID, <span class="sc">-</span>YORF, <span class="sc">-</span>somenumber, <span class="sc">-</span>GWEIGHT) <span class="sc">|&gt;</span> </span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gather</span>(<span class="at">key=</span>nutrientrate, <span class="at">value=</span>expression, G0<span class="fl">.05</span><span class="sc">:</span>U0<span class="fl">.3</span>) <span class="sc">|&gt;</span> </span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">separate</span>(nutrientrate, <span class="at">into=</span><span class="fu">c</span>(<span class="st">"nutrient"</span>, <span class="st">"rate"</span>), <span class="at">sep=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="inner_join-to-go" class="level3" data-number="4.3.3">
<h3 data-number="4.3.3" class="anchored" data-anchor-id="inner_join-to-go"><span class="header-section-number">4.3.3</span> <code>inner_join()</code> to GO</h3>
<p>It’s rare that a data analysis involves only a single table of data. You normally have many tables that contribute to an analysis, and you need flexible tools to combine them. The <strong>dplyr</strong> package has several tools that let you work with multiple tables at once. Do a <a href="https://www.google.com/search?q=SQL+join&amp;tbm=isch">Google image search for “SQL Joins”</a>, and look at <a href="http://www.rstudio.com/wp-content/uploads/2015/02/data-wrangling-cheatsheet.pdf">RStudio’s Data Wrangling Cheat Sheet</a> to learn more.</p>
<p>First, let’s import the dataset that links the systematic name to gene ontology information. It’s the <a href="data/brauer2007_sysname2go.csv">brauer2007_sysname2go.csv</a> file. Let’s call the imported data frame <code>sn2go</code>.</p>
<div class="cell" data-hash="tidyr_cache/html/unnamed-chunk-8_27bb33d33569cb609bee672bf224d697">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Import the data</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>sn2go <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">"data/brauer2007_sysname2go.csv"</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a look</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co"># View(sn2go)</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sn2go)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  systematic_name bp                           mf                           
  &lt;chr&gt;           &lt;chr&gt;                        &lt;chr&gt;                        
1 YNL049C         ER to Golgi transport        molecular function unknown   
2 YNL095C         biological process unknown   molecular function unknown   
3 YDL104C         proteolysis and peptidolysis metalloendopeptidase activity
4 YLR115W         mRNA polyadenylylation*      RNA binding                  
5 YMR183C         vesicle fusion*              t-SNARE activity             
6 YML017W         biological process unknown   molecular function unknown   </code></pre>
</div>
</div>
<p>Now, look up some help for <code>?inner_join</code>. Inner join will return a table with all rows from the first table where there are matching rows in the second table, and returns all columns from both tables. Let’s give this a try.</p>
<div class="cell" data-hash="tidyr_cache/html/innerjoin_7f1e4cf8f9c9ec0121b2b82f727b72a1">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>yjoined <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(ynogo, sn2go, <span class="at">by=</span><span class="st">"systematic_name"</span>)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="co"># View(yjoined)</span></span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>yjoined</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 199,296 × 7
   symbol systematic_name nutrient rate  expression bp                     mf   
   &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;    &lt;chr&gt;      &lt;dbl&gt; &lt;chr&gt;                  &lt;chr&gt;
 1 SFB2   YNL049C         G        0.05       -0.24 ER to Golgi transport  mole…
 2 NA     YNL095C         G        0.05        0.28 biological process un… mole…
 3 QRI7   YDL104C         G        0.05       -0.02 proteolysis and pepti… meta…
 4 CFT2   YLR115W         G        0.05       -0.33 mRNA polyadenylylatio… RNA …
 5 SSO2   YMR183C         G        0.05        0.05 vesicle fusion*        t-SN…
 6 PSP2   YML017W         G        0.05       -0.69 biological process un… mole…
 7 RIB2   YOL066C         G        0.05       -0.55 riboflavin biosynthes… pseu…
 8 VMA13  YPR036W         G        0.05       -0.75 vacuolar acidification hydr…
 9 EDC3   YEL015W         G        0.05       -0.24 deadenylylation-indep… mole…
10 VPS5   YOR069W         G        0.05       -0.16 protein retention in … prot…
# ℹ 199,286 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># The glimpse function makes it possible to see a little bit of everything in your data.</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glimpse</span>(yjoined)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Rows: 199,296
Columns: 7
$ symbol          &lt;chr&gt; "SFB2", "NA", "QRI7", "CFT2", "SSO2", "PSP2", "RIB2", …
$ systematic_name &lt;chr&gt; "YNL049C", "YNL095C", "YDL104C", "YLR115W", "YMR183C",…
$ nutrient        &lt;chr&gt; "G", "G", "G", "G", "G", "G", "G", "G", "G", "G", "G",…
$ rate            &lt;chr&gt; "0.05", "0.05", "0.05", "0.05", "0.05", "0.05", "0.05"…
$ expression      &lt;dbl&gt; -0.24, 0.28, -0.02, -0.33, 0.05, -0.69, -0.55, -0.75, …
$ bp              &lt;chr&gt; "ER to Golgi transport", "biological process unknown",…
$ mf              &lt;chr&gt; "molecular function unknown", "molecular function unkn…</code></pre>
</div>
</div>
<p>There are many different kinds of two-table verbs/joins in dplyr. In this example, every systematic name in <code>ynogo</code> had a corresponding entry in <code>sn2go</code>, but if this weren’t the case, those un-annotated genes would have been removed entirely by the <code>inner_join</code>. A <code>left_join</code> would have returned all the rows in <code>ynogo</code>, but would have filled in <code>bp</code> and <code>mf</code> with missing values (<code>NA</code>) when there was no corresponding entry. See also: <code>right_join</code>, <code>semi_join</code>, and <code>anti_join</code>.</p>
</section>
<section id="finishing-touches" class="level3" data-number="4.3.4">
<h3 data-number="4.3.4" class="anchored" data-anchor-id="finishing-touches"><span class="header-section-number">4.3.4</span> Finishing touches</h3>
<p>We’re almost there but we have an obvious discrepancy in the number of rows between <code>yjoined</code> and <code>ydat</code>.</p>
<div class="cell" data-hash="tidyr_cache/html/rowcheck_8299de7ebceb98b3dc1eab7494bdaaec">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(yjoined)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 199296</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(ydat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 198430</code></pre>
</div>
</div>
<p>Before we can figure out what rows are different, we need to make sure all of the columns are the same class and do something more miscellaneous cleanup.</p>
<p>In particular:</p>
<ol type="1">
<li>Convert rate to a numeric column</li>
<li>Make sure <code>NA</code> values are coded properly</li>
<li>Create (and merge) values to convert “G” to “Glucose”, “L” to “Leucine”, etc.</li>
<li>Rename and reorder columns</li>
</ol>
<p>The code below implements those operations on <code>yjoined</code>.</p>
<div class="cell" data-hash="tidyr_cache/html/misctidy_866bfd1f2184d3a249f8b7f64adfcceb">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>nutrientlookup <span class="ot">&lt;-</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">nutrient =</span> <span class="fu">c</span>(<span class="st">"G"</span>, <span class="st">"L"</span>, <span class="st">"N"</span>, <span class="st">"P"</span>, <span class="st">"S"</span>, <span class="st">"U"</span>), <span class="at">nutrientname =</span> <span class="fu">c</span>(<span class="st">"Glucose"</span>, <span class="st">"Leucine"</span>, <span class="st">"Ammonia"</span>,<span class="st">"Phosphate"</span>, <span class="st">"Sulfate"</span>,<span class="st">"Uracil"</span>))</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>yjoined <span class="ot">&lt;-</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  yjoined <span class="sc">|&gt;</span></span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">rate =</span> <span class="fu">as.numeric</span>(rate)) <span class="sc">|&gt;</span></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">symbol =</span> <span class="fu">ifelse</span>(symbol <span class="sc">==</span> <span class="st">"NA"</span>, <span class="cn">NA</span>, symbol)) <span class="sc">|&gt;</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">left_join</span>(nutrientlookup) <span class="sc">|&gt;</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(<span class="sc">-</span>nutrient) <span class="sc">|&gt;</span></span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(symbol<span class="sc">:</span>systematic_name, <span class="at">nutrient =</span> nutrientname, rate<span class="sc">:</span>mf)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now we can determine what rows are different between <code>yjoined</code> and <code>ydat</code> using <code>anti_join</code>, which will return all of the rows that <em>do not</em> match.</p>
<div class="cell" data-hash="tidyr_cache/html/antijoin_36c54890b4607b8433063125f94785c2">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu">anti_join</span>(yjoined, ydat) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 866 × 7
   symbol systematic_name nutrient  rate expression bp                     mf   
   &lt;chr&gt;  &lt;chr&gt;           &lt;chr&gt;    &lt;dbl&gt;      &lt;dbl&gt; &lt;chr&gt;                  &lt;chr&gt;
 1 &lt;NA&gt;   YLL030C         Glucose   0.05         NA &lt;NA&gt;                   &lt;NA&gt; 
 2 &lt;NA&gt;   YOR050C         Glucose   0.05         NA &lt;NA&gt;                   &lt;NA&gt; 
 3 &lt;NA&gt;   YPR039W         Glucose   0.05         NA &lt;NA&gt;                   &lt;NA&gt; 
 4 &lt;NA&gt;   YOL013W-B       Glucose   0.05         NA &lt;NA&gt;                   &lt;NA&gt; 
 5 HXT12  YIL170W         Glucose   0.05         NA biological process un… mole…
 6 &lt;NA&gt;   YPR013C         Glucose   0.05         NA biological process un… mole…
 7 &lt;NA&gt;   YOR314W         Glucose   0.05         NA &lt;NA&gt;                   &lt;NA&gt; 
 8 &lt;NA&gt;   YJL064W         Glucose   0.05         NA &lt;NA&gt;                   &lt;NA&gt; 
 9 &lt;NA&gt;   YPR136C         Glucose   0.05         NA &lt;NA&gt;                   &lt;NA&gt; 
10 &lt;NA&gt;   YDR015C         Glucose   0.05         NA &lt;NA&gt;                   &lt;NA&gt; 
# ℹ 856 more rows</code></pre>
</div>
</div>
<p>Hmmmm … so <code>yjoined</code> has some rows that have missing (<code>NA</code>) expression values. Let’s try removing those and then comparing the data frame contents one more time.</p>
<div class="cell" data-hash="tidyr_cache/html/rowcheck2_21fa8adc0713f5c0a9023ca727f8ab55">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>yjoined <span class="ot">&lt;-</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  yjoined <span class="sc">|&gt;</span></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">is.na</span>(expression))</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(yjoined)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 198430</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(ydat)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 198430</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(ydat, yjoined)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Attributes: &lt; Names: 1 string mismatch &gt;"                                              
[2] "Attributes: &lt; Length mismatch: comparison on first 2 components &gt;"                     
[3] "Attributes: &lt; Component \"class\": Lengths (4, 3) differ (string compare on first 3) &gt;"
[4] "Attributes: &lt; Component \"class\": 3 string mismatches &gt;"                              
[5] "Attributes: &lt; Component 2: target is externalptr, current is numeric &gt;"                </code></pre>
</div>
</div>
<p>Looks like that did it!</p>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./dplyr.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Data Manipulation with dplyr</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./ggplot2.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Visualization with ggplot2</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">© 2018 Stephen D. Turner</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
      <ul class="footer-items list-unstyled">
    <li class="nav-item compact">
    <a class="nav-link" href="https://x.com/strnr">
      <i class="bi bi-twitter" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/stephenturner">
      <i class="bi bi-github" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/in/turnersd/">
      <i class="bi bi-linkedin" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://stephenturner.us/">
      <i class="bi bi-house-fill" role="img">
</i> 
    </a>
  </li>  
    <li class="nav-item compact">
    <a class="nav-link" href="https://blog.stephenturner.us/">
      <i class="bi bi-rss" role="img">
</i> 
    </a>
  </li>  
</ul>
    </div>
  </div>
</footer>



</body></html>